
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="A federated learning implementation of DESeq2 for differential expression analysis.">
      
      
        <meta name="author" content="Boris Muzellec, Ulysse Marteau-Ferey, Tangy Marchand">
      
      
        <link rel="canonical" href="https://github.com/owkin/fedpydeseq2/api/core/deseq2_core/deseq2_core/">
      
      
        <link rel="prev" href="../../../fedpydeseq2_pipeline/">
      
      
        <link rel="next" href="../build_design_matrix/build_design_matrix/">
      
      
      <link rel="icon" href="../../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.12">
    
    
      
        <title>Home - FedPyDESeq2</title>
      
    
    
      <link rel="stylesheet" href="../../../../assets/stylesheets/main.2afb09e1.min.css">
      
        
        <link rel="stylesheet" href="../../../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../../assets/_mkdocstrings.css">
    
      <link rel="stylesheet" href="../../../../css/timeago.css">
    
      <link rel="stylesheet" href="https://unpkg.com/katex@0/dist/katex.min.css">
    
      <link rel="stylesheet" href="../../../../sg_gallery-rendered-html.css">
    
      <link rel="stylesheet" href="../../../../sg_gallery-binder.css">
    
      <link rel="stylesheet" href="../../../../sg_gallery.css">
    
      <link rel="stylesheet" href="../../../../sg_gallery-dataframe.css">
    
    <script>__md_scope=new URL("../../../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#core-of-the-fed-pydeseq2-package-deseq2_core" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../../.." title="FedPyDESeq2" class="md-header__button md-logo" aria-label="FedPyDESeq2" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            FedPyDESeq2
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Home
            
          </span>
        </div>
      </div>
    </div>
    
      
    
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/owkin/fedpydeseq2" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    owkin/fedpydeseq2
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../../.." title="FedPyDESeq2" class="md-nav__button md-logo" aria-label="FedPyDESeq2" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    FedPyDESeq2
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/owkin/fedpydeseq2" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    owkin/fedpydeseq2
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    
    
      
        
          
        
      
        
      
        
      
        
      
        
      
        
      
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_1" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../.." class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    General
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_1" id="__nav_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_1">
            <span class="md-nav__icon md-icon"></span>
            General
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../usage/installation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Installation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../usage/installation_developers/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Installation for developers
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../usage/contributing/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Contributing
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../usage/logging/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Logging
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../usage/references/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    References
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
          
        
      
        
      
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" >
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../../generated/gallery/" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    Tutorials
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_2" id="__nav_2_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Tutorials
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../generated/gallery/plot_demo/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    FedPyDESeq2 demo on the TCGA-LUAD dataset.
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
      
        
          
        
      
        
      
        
      
        
      
        
      
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" checked>
        
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../" class="md-nav__link ">
              
  
  
  <span class="md-ellipsis">
    API
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_3" id="__nav_3_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            API
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../fedpydeseq2_pipeline/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Running an experiment
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_3" checked>
        
          
          <label class="md-nav__link" for="__nav_3_3" id="__nav_3_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Core
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_3_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_3_3">
            <span class="md-nav__icon md-icon"></span>
            Core
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#workflow-graph" class="md-nav__link">
    <span class="md-ellipsis">
      Workflow graph
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../build_design_matrix/build_design_matrix/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Build design matrix
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../compute_size_factors/compute_size_factors/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Compute size factors
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../deseq2_lfc_dispersions/deseq2_lfc_dispersions/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Estimating LFCs and dispersions
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../compute_cook_distance/compute_cook_distance/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Computing the Cooks distance
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../replace_outliers/replace_outliers/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Replacing outliers
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../replace_refitted_values/replace_refitted_values/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Replace refitted values
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../deseq2_stats/deseq2_stats/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Computing statistics and p-values
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../save_pipeline_results/save_pipeline_results/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Saving results
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../strategy/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Strategy
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_4" >
        
          
          <label class="md-nav__link" for="__nav_3_4" id="__nav_3_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Utils
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_4">
            <span class="md-nav__icon md-icon"></span>
            Utils
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../utils/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Core
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../substra_utils/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Substra
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_5" >
        
          
          <label class="md-nav__link" for="__nav_3_5" id="__nav_3_5_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Federated algorithms
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_5">
            <span class="md-nav__icon md-icon"></span>
            Federated algorithms
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../federated_algorithms/federated_algorithms/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../federated_algorithms/compute_trimmed_mean/compute_trimmed_mean/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Compute trimmed mean
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../federated_algorithms/dispersions_grid_search/dispersions_grid_search/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Compute dispersions with grid search
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../federated_algorithms/fed_irls/fed_irls/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Federated IRLS
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../federated_algorithms/fed_pqn/fed_pqn/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Federated ProxQuasiNewton
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#workflow-graph" class="md-nav__link">
    <span class="md-ellipsis">
      Workflow graph
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="core-of-the-fed-pydeseq2-package-deseq2_core">Core of the fed-pydeseq2 package: <code>deseq2_core</code></h1>
<h2 id="workflow-graph">Workflow graph</h2>
<p>The workflow graph below illustrates the sequence of operations in the design matrix construction process. It shows how data flows between local centers and the aggregation server during the <code>run_deseq_pipe</code> function.</p>
<p><img alt="Workflow Graph" src="../workflow_graph.png" /></p>
<p>For a detailed breakdown of the shared states and their contents at each step, please refer to the table below.</p>
<h1 id="api">API</h1>
<p>The core of the fed-pydeseq2 package is encapsulated in the following Mixin class.</p>


<div class="doc doc-object doc-module">



<a id="fedpydeseq2.core.deseq2_core.deseq2_full_pipe"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">








<div class="doc doc-object doc-class">



<h2 id="fedpydeseq2.core.deseq2_core.deseq2_full_pipe.DESeq2FullPipe" class="doc doc-heading">
            <code>DESeq2FullPipe</code>


</h2>


    <div class="doc doc-contents ">
            <p class="doc doc-class-bases">
              Bases: <code><a class="autorefs autorefs-internal" title="BuildDesignMatrix (fedpydeseq2.core.deseq2_core.build_design_matrix.BuildDesignMatrix)" href="../build_design_matrix/build_design_matrix/#fedpydeseq2.core.deseq2_core.build_design_matrix.build_design_matrix.BuildDesignMatrix">BuildDesignMatrix</a></code>, <code><a class="autorefs autorefs-internal" title="ComputeSizeFactors (fedpydeseq2.core.deseq2_core.compute_size_factors.ComputeSizeFactors)" href="../compute_size_factors/compute_size_factors/#fedpydeseq2.core.deseq2_core.compute_size_factors.compute_size_factors.ComputeSizeFactors">ComputeSizeFactors</a></code>, <code><a class="autorefs autorefs-internal" title="DESeq2LFCDispersions (fedpydeseq2.core.deseq2_core.deseq2_lfc_dispersions.DESeq2LFCDispersions)" href="../deseq2_lfc_dispersions/deseq2_lfc_dispersions/#fedpydeseq2.core.deseq2_core.deseq2_lfc_dispersions.deseq2_lfc_dispersions.DESeq2LFCDispersions">DESeq2LFCDispersions</a></code>, <code><a class="autorefs autorefs-internal" title="ComputeCookDistances (fedpydeseq2.core.deseq2_core.compute_cook_distance.ComputeCookDistances)" href="../compute_cook_distance/compute_cook_distance/#fedpydeseq2.core.deseq2_core.compute_cook_distance.compute_cook_distance.ComputeCookDistances">ComputeCookDistances</a></code>, <code><a class="autorefs autorefs-internal" title="ReplaceCooksOutliers (fedpydeseq2.core.deseq2_core.replace_outliers.ReplaceCooksOutliers)" href="../replace_outliers/replace_outliers/#fedpydeseq2.core.deseq2_core.replace_outliers.replace_outliers.ReplaceCooksOutliers">ReplaceCooksOutliers</a></code>, <code><a class="autorefs autorefs-internal" title="ReplaceRefittedValues (fedpydeseq2.core.deseq2_core.replace_refitted_values.ReplaceRefittedValues)" href="../replace_refitted_values/replace_refitted_values/#fedpydeseq2.core.deseq2_core.replace_refitted_values.replace_refitted_values.ReplaceRefittedValues">ReplaceRefittedValues</a></code>, <code><a class="autorefs autorefs-internal" title="DESeq2Stats (fedpydeseq2.core.deseq2_core.deseq2_stats.deseq2_stats.DESeq2Stats)" href="../deseq2_stats/deseq2_stats/#fedpydeseq2.core.deseq2_core.deseq2_stats.deseq2_stats.DESeq2Stats">DESeq2Stats</a></code>, <code><a class="autorefs autorefs-internal" title="SavePipelineResults (fedpydeseq2.core.deseq2_core.save_pipeline_results.SavePipelineResults)" href="../save_pipeline_results/save_pipeline_results/#fedpydeseq2.core.deseq2_core.save_pipeline_results.SavePipelineResults">SavePipelineResults</a></code></p>


        <p>A Mixin class to run the full DESeq2 pipeline.</p>


<p><span class="doc-section-title">Methods:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
            <tr class="doc-section-item">
              <td><code><a class="autorefs autorefs-internal" title="run_deseq_pipe(train_data_nodes, aggregation_node, local_states, round_idx=0, clean_models=True, clean_last_model=False) (fedpydeseq2.core.deseq2_core.deseq2_full_pipe.DESeq2FullPipe.run_deseq_pipe)" href="#fedpydeseq2.core.deseq2_core.deseq2_full_pipe.DESeq2FullPipe.run_deseq_pipe">run_deseq_pipe</a></code></td>
              <td>
                <div class="doc-md-description">
                  <p>The method to run the full DESeq2 pipeline.</p>
                </div>
              </td>
            </tr>
      </tbody>
    </table>







              <details class="quote">
                <summary>Source code in <code>fedpydeseq2/core/deseq2_core/deseq2_full_pipe.py</code></summary>
                <div class="language-python highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-17"> 17</a></span>
<span class="normal"><a href="#__codelineno-0-18"> 18</a></span>
<span class="normal"><a href="#__codelineno-0-19"> 19</a></span>
<span class="normal"><a href="#__codelineno-0-20"> 20</a></span>
<span class="normal"><a href="#__codelineno-0-21"> 21</a></span>
<span class="normal"><a href="#__codelineno-0-22"> 22</a></span>
<span class="normal"><a href="#__codelineno-0-23"> 23</a></span>
<span class="normal"><a href="#__codelineno-0-24"> 24</a></span>
<span class="normal"><a href="#__codelineno-0-25"> 25</a></span>
<span class="normal"><a href="#__codelineno-0-26"> 26</a></span>
<span class="normal"><a href="#__codelineno-0-27"> 27</a></span>
<span class="normal"><a href="#__codelineno-0-28"> 28</a></span>
<span class="normal"><a href="#__codelineno-0-29"> 29</a></span>
<span class="normal"><a href="#__codelineno-0-30"> 30</a></span>
<span class="normal"><a href="#__codelineno-0-31"> 31</a></span>
<span class="normal"><a href="#__codelineno-0-32"> 32</a></span>
<span class="normal"><a href="#__codelineno-0-33"> 33</a></span>
<span class="normal"><a href="#__codelineno-0-34"> 34</a></span>
<span class="normal"><a href="#__codelineno-0-35"> 35</a></span>
<span class="normal"><a href="#__codelineno-0-36"> 36</a></span>
<span class="normal"><a href="#__codelineno-0-37"> 37</a></span>
<span class="normal"><a href="#__codelineno-0-38"> 38</a></span>
<span class="normal"><a href="#__codelineno-0-39"> 39</a></span>
<span class="normal"><a href="#__codelineno-0-40"> 40</a></span>
<span class="normal"><a href="#__codelineno-0-41"> 41</a></span>
<span class="normal"><a href="#__codelineno-0-42"> 42</a></span>
<span class="normal"><a href="#__codelineno-0-43"> 43</a></span>
<span class="normal"><a href="#__codelineno-0-44"> 44</a></span>
<span class="normal"><a href="#__codelineno-0-45"> 45</a></span>
<span class="normal"><a href="#__codelineno-0-46"> 46</a></span>
<span class="normal"><a href="#__codelineno-0-47"> 47</a></span>
<span class="normal"><a href="#__codelineno-0-48"> 48</a></span>
<span class="normal"><a href="#__codelineno-0-49"> 49</a></span>
<span class="normal"><a href="#__codelineno-0-50"> 50</a></span>
<span class="normal"><a href="#__codelineno-0-51"> 51</a></span>
<span class="normal"><a href="#__codelineno-0-52"> 52</a></span>
<span class="normal"><a href="#__codelineno-0-53"> 53</a></span>
<span class="normal"><a href="#__codelineno-0-54"> 54</a></span>
<span class="normal"><a href="#__codelineno-0-55"> 55</a></span>
<span class="normal"><a href="#__codelineno-0-56"> 56</a></span>
<span class="normal"><a href="#__codelineno-0-57"> 57</a></span>
<span class="normal"><a href="#__codelineno-0-58"> 58</a></span>
<span class="normal"><a href="#__codelineno-0-59"> 59</a></span>
<span class="normal"><a href="#__codelineno-0-60"> 60</a></span>
<span class="normal"><a href="#__codelineno-0-61"> 61</a></span>
<span class="normal"><a href="#__codelineno-0-62"> 62</a></span>
<span class="normal"><a href="#__codelineno-0-63"> 63</a></span>
<span class="normal"><a href="#__codelineno-0-64"> 64</a></span>
<span class="normal"><a href="#__codelineno-0-65"> 65</a></span>
<span class="normal"><a href="#__codelineno-0-66"> 66</a></span>
<span class="normal"><a href="#__codelineno-0-67"> 67</a></span>
<span class="normal"><a href="#__codelineno-0-68"> 68</a></span>
<span class="normal"><a href="#__codelineno-0-69"> 69</a></span>
<span class="normal"><a href="#__codelineno-0-70"> 70</a></span>
<span class="normal"><a href="#__codelineno-0-71"> 71</a></span>
<span class="normal"><a href="#__codelineno-0-72"> 72</a></span>
<span class="normal"><a href="#__codelineno-0-73"> 73</a></span>
<span class="normal"><a href="#__codelineno-0-74"> 74</a></span>
<span class="normal"><a href="#__codelineno-0-75"> 75</a></span>
<span class="normal"><a href="#__codelineno-0-76"> 76</a></span>
<span class="normal"><a href="#__codelineno-0-77"> 77</a></span>
<span class="normal"><a href="#__codelineno-0-78"> 78</a></span>
<span class="normal"><a href="#__codelineno-0-79"> 79</a></span>
<span class="normal"><a href="#__codelineno-0-80"> 80</a></span>
<span class="normal"><a href="#__codelineno-0-81"> 81</a></span>
<span class="normal"><a href="#__codelineno-0-82"> 82</a></span>
<span class="normal"><a href="#__codelineno-0-83"> 83</a></span>
<span class="normal"><a href="#__codelineno-0-84"> 84</a></span>
<span class="normal"><a href="#__codelineno-0-85"> 85</a></span>
<span class="normal"><a href="#__codelineno-0-86"> 86</a></span>
<span class="normal"><a href="#__codelineno-0-87"> 87</a></span>
<span class="normal"><a href="#__codelineno-0-88"> 88</a></span>
<span class="normal"><a href="#__codelineno-0-89"> 89</a></span>
<span class="normal"><a href="#__codelineno-0-90"> 90</a></span>
<span class="normal"><a href="#__codelineno-0-91"> 91</a></span>
<span class="normal"><a href="#__codelineno-0-92"> 92</a></span>
<span class="normal"><a href="#__codelineno-0-93"> 93</a></span>
<span class="normal"><a href="#__codelineno-0-94"> 94</a></span>
<span class="normal"><a href="#__codelineno-0-95"> 95</a></span>
<span class="normal"><a href="#__codelineno-0-96"> 96</a></span>
<span class="normal"><a href="#__codelineno-0-97"> 97</a></span>
<span class="normal"><a href="#__codelineno-0-98"> 98</a></span>
<span class="normal"><a href="#__codelineno-0-99"> 99</a></span>
<span class="normal"><a href="#__codelineno-0-100">100</a></span>
<span class="normal"><a href="#__codelineno-0-101">101</a></span>
<span class="normal"><a href="#__codelineno-0-102">102</a></span>
<span class="normal"><a href="#__codelineno-0-103">103</a></span>
<span class="normal"><a href="#__codelineno-0-104">104</a></span>
<span class="normal"><a href="#__codelineno-0-105">105</a></span>
<span class="normal"><a href="#__codelineno-0-106">106</a></span>
<span class="normal"><a href="#__codelineno-0-107">107</a></span>
<span class="normal"><a href="#__codelineno-0-108">108</a></span>
<span class="normal"><a href="#__codelineno-0-109">109</a></span>
<span class="normal"><a href="#__codelineno-0-110">110</a></span>
<span class="normal"><a href="#__codelineno-0-111">111</a></span>
<span class="normal"><a href="#__codelineno-0-112">112</a></span>
<span class="normal"><a href="#__codelineno-0-113">113</a></span>
<span class="normal"><a href="#__codelineno-0-114">114</a></span>
<span class="normal"><a href="#__codelineno-0-115">115</a></span>
<span class="normal"><a href="#__codelineno-0-116">116</a></span>
<span class="normal"><a href="#__codelineno-0-117">117</a></span>
<span class="normal"><a href="#__codelineno-0-118">118</a></span>
<span class="normal"><a href="#__codelineno-0-119">119</a></span>
<span class="normal"><a href="#__codelineno-0-120">120</a></span>
<span class="normal"><a href="#__codelineno-0-121">121</a></span>
<span class="normal"><a href="#__codelineno-0-122">122</a></span>
<span class="normal"><a href="#__codelineno-0-123">123</a></span>
<span class="normal"><a href="#__codelineno-0-124">124</a></span>
<span class="normal"><a href="#__codelineno-0-125">125</a></span>
<span class="normal"><a href="#__codelineno-0-126">126</a></span>
<span class="normal"><a href="#__codelineno-0-127">127</a></span>
<span class="normal"><a href="#__codelineno-0-128">128</a></span>
<span class="normal"><a href="#__codelineno-0-129">129</a></span>
<span class="normal"><a href="#__codelineno-0-130">130</a></span>
<span class="normal"><a href="#__codelineno-0-131">131</a></span>
<span class="normal"><a href="#__codelineno-0-132">132</a></span>
<span class="normal"><a href="#__codelineno-0-133">133</a></span>
<span class="normal"><a href="#__codelineno-0-134">134</a></span>
<span class="normal"><a href="#__codelineno-0-135">135</a></span>
<span class="normal"><a href="#__codelineno-0-136">136</a></span>
<span class="normal"><a href="#__codelineno-0-137">137</a></span>
<span class="normal"><a href="#__codelineno-0-138">138</a></span>
<span class="normal"><a href="#__codelineno-0-139">139</a></span>
<span class="normal"><a href="#__codelineno-0-140">140</a></span>
<span class="normal"><a href="#__codelineno-0-141">141</a></span>
<span class="normal"><a href="#__codelineno-0-142">142</a></span>
<span class="normal"><a href="#__codelineno-0-143">143</a></span>
<span class="normal"><a href="#__codelineno-0-144">144</a></span>
<span class="normal"><a href="#__codelineno-0-145">145</a></span>
<span class="normal"><a href="#__codelineno-0-146">146</a></span>
<span class="normal"><a href="#__codelineno-0-147">147</a></span>
<span class="normal"><a href="#__codelineno-0-148">148</a></span>
<span class="normal"><a href="#__codelineno-0-149">149</a></span>
<span class="normal"><a href="#__codelineno-0-150">150</a></span>
<span class="normal"><a href="#__codelineno-0-151">151</a></span>
<span class="normal"><a href="#__codelineno-0-152">152</a></span>
<span class="normal"><a href="#__codelineno-0-153">153</a></span>
<span class="normal"><a href="#__codelineno-0-154">154</a></span>
<span class="normal"><a href="#__codelineno-0-155">155</a></span>
<span class="normal"><a href="#__codelineno-0-156">156</a></span>
<span class="normal"><a href="#__codelineno-0-157">157</a></span>
<span class="normal"><a href="#__codelineno-0-158">158</a></span>
<span class="normal"><a href="#__codelineno-0-159">159</a></span>
<span class="normal"><a href="#__codelineno-0-160">160</a></span>
<span class="normal"><a href="#__codelineno-0-161">161</a></span>
<span class="normal"><a href="#__codelineno-0-162">162</a></span>
<span class="normal"><a href="#__codelineno-0-163">163</a></span>
<span class="normal"><a href="#__codelineno-0-164">164</a></span>
<span class="normal"><a href="#__codelineno-0-165">165</a></span>
<span class="normal"><a href="#__codelineno-0-166">166</a></span>
<span class="normal"><a href="#__codelineno-0-167">167</a></span>
<span class="normal"><a href="#__codelineno-0-168">168</a></span>
<span class="normal"><a href="#__codelineno-0-169">169</a></span>
<span class="normal"><a href="#__codelineno-0-170">170</a></span>
<span class="normal"><a href="#__codelineno-0-171">171</a></span>
<span class="normal"><a href="#__codelineno-0-172">172</a></span>
<span class="normal"><a href="#__codelineno-0-173">173</a></span>
<span class="normal"><a href="#__codelineno-0-174">174</a></span>
<span class="normal"><a href="#__codelineno-0-175">175</a></span>
<span class="normal"><a href="#__codelineno-0-176">176</a></span>
<span class="normal"><a href="#__codelineno-0-177">177</a></span>
<span class="normal"><a href="#__codelineno-0-178">178</a></span>
<span class="normal"><a href="#__codelineno-0-179">179</a></span>
<span class="normal"><a href="#__codelineno-0-180">180</a></span>
<span class="normal"><a href="#__codelineno-0-181">181</a></span>
<span class="normal"><a href="#__codelineno-0-182">182</a></span>
<span class="normal"><a href="#__codelineno-0-183">183</a></span>
<span class="normal"><a href="#__codelineno-0-184">184</a></span>
<span class="normal"><a href="#__codelineno-0-185">185</a></span>
<span class="normal"><a href="#__codelineno-0-186">186</a></span>
<span class="normal"><a href="#__codelineno-0-187">187</a></span>
<span class="normal"><a href="#__codelineno-0-188">188</a></span>
<span class="normal"><a href="#__codelineno-0-189">189</a></span>
<span class="normal"><a href="#__codelineno-0-190">190</a></span>
<span class="normal"><a href="#__codelineno-0-191">191</a></span></pre></div></td><td class="code"><div><pre><span></span><code><span id="__span-0-17"><a id="__codelineno-0-17" name="__codelineno-0-17"></a><span class="k">class</span><span class="w"> </span><span class="nc">DESeq2FullPipe</span><span class="p">(</span>
</span><span id="__span-0-18"><a id="__codelineno-0-18" name="__codelineno-0-18"></a>    <span class="n">BuildDesignMatrix</span><span class="p">,</span>
</span><span id="__span-0-19"><a id="__codelineno-0-19" name="__codelineno-0-19"></a>    <span class="n">ComputeSizeFactors</span><span class="p">,</span>
</span><span id="__span-0-20"><a id="__codelineno-0-20" name="__codelineno-0-20"></a>    <span class="n">DESeq2LFCDispersions</span><span class="p">,</span>
</span><span id="__span-0-21"><a id="__codelineno-0-21" name="__codelineno-0-21"></a>    <span class="n">ComputeCookDistances</span><span class="p">,</span>
</span><span id="__span-0-22"><a id="__codelineno-0-22" name="__codelineno-0-22"></a>    <span class="n">ReplaceCooksOutliers</span><span class="p">,</span>
</span><span id="__span-0-23"><a id="__codelineno-0-23" name="__codelineno-0-23"></a>    <span class="n">ReplaceRefittedValues</span><span class="p">,</span>
</span><span id="__span-0-24"><a id="__codelineno-0-24" name="__codelineno-0-24"></a>    <span class="n">DESeq2Stats</span><span class="p">,</span>
</span><span id="__span-0-25"><a id="__codelineno-0-25" name="__codelineno-0-25"></a>    <span class="n">SavePipelineResults</span><span class="p">,</span>
</span><span id="__span-0-26"><a id="__codelineno-0-26" name="__codelineno-0-26"></a><span class="p">):</span>
</span><span id="__span-0-27"><a id="__codelineno-0-27" name="__codelineno-0-27"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;A Mixin class to run the full DESeq2 pipeline.</span>
</span><span id="__span-0-28"><a id="__codelineno-0-28" name="__codelineno-0-28"></a>
</span><span id="__span-0-29"><a id="__codelineno-0-29" name="__codelineno-0-29"></a><span class="sd">    Methods</span>
</span><span id="__span-0-30"><a id="__codelineno-0-30" name="__codelineno-0-30"></a><span class="sd">    -------</span>
</span><span id="__span-0-31"><a id="__codelineno-0-31" name="__codelineno-0-31"></a><span class="sd">    run_deseq_pipe</span>
</span><span id="__span-0-32"><a id="__codelineno-0-32" name="__codelineno-0-32"></a><span class="sd">        The method to run the full DESeq2 pipeline.</span>
</span><span id="__span-0-33"><a id="__codelineno-0-33" name="__codelineno-0-33"></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="__span-0-34"><a id="__codelineno-0-34" name="__codelineno-0-34"></a>
</span><span id="__span-0-35"><a id="__codelineno-0-35" name="__codelineno-0-35"></a>    <span class="nd">@log_organisation_method</span>
</span><span id="__span-0-36"><a id="__codelineno-0-36" name="__codelineno-0-36"></a>    <span class="k">def</span><span class="w"> </span><span class="nf">run_deseq_pipe</span><span class="p">(</span>
</span><span id="__span-0-37"><a id="__codelineno-0-37" name="__codelineno-0-37"></a>        <span class="bp">self</span><span class="p">,</span>
</span><span id="__span-0-38"><a id="__codelineno-0-38" name="__codelineno-0-38"></a>        <span class="n">train_data_nodes</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="n">TrainDataNode</span><span class="p">],</span>
</span><span id="__span-0-39"><a id="__codelineno-0-39" name="__codelineno-0-39"></a>        <span class="n">aggregation_node</span><span class="p">:</span> <span class="n">AggregationNode</span><span class="p">,</span>
</span><span id="__span-0-40"><a id="__codelineno-0-40" name="__codelineno-0-40"></a>        <span class="n">local_states</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">LocalStateRef</span><span class="p">],</span>
</span><span id="__span-0-41"><a id="__codelineno-0-41" name="__codelineno-0-41"></a>        <span class="n">round_idx</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>
</span><span id="__span-0-42"><a id="__codelineno-0-42" name="__codelineno-0-42"></a>        <span class="n">clean_models</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
</span><span id="__span-0-43"><a id="__codelineno-0-43" name="__codelineno-0-43"></a>        <span class="n">clean_last_model</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
</span><span id="__span-0-44"><a id="__codelineno-0-44" name="__codelineno-0-44"></a>    <span class="p">):</span>
</span><span id="__span-0-45"><a id="__codelineno-0-45" name="__codelineno-0-45"></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;Run the DESeq2 pipeline.</span>
</span><span id="__span-0-46"><a id="__codelineno-0-46" name="__codelineno-0-46"></a>
</span><span id="__span-0-47"><a id="__codelineno-0-47" name="__codelineno-0-47"></a><span class="sd">        Parameters</span>
</span><span id="__span-0-48"><a id="__codelineno-0-48" name="__codelineno-0-48"></a><span class="sd">        ----------</span>
</span><span id="__span-0-49"><a id="__codelineno-0-49" name="__codelineno-0-49"></a><span class="sd">        train_data_nodes : list[TrainDataNode]</span>
</span><span id="__span-0-50"><a id="__codelineno-0-50" name="__codelineno-0-50"></a><span class="sd">            List of the train nodes.</span>
</span><span id="__span-0-51"><a id="__codelineno-0-51" name="__codelineno-0-51"></a><span class="sd">        aggregation_node : AggregationNode</span>
</span><span id="__span-0-52"><a id="__codelineno-0-52" name="__codelineno-0-52"></a><span class="sd">            Aggregation node.</span>
</span><span id="__span-0-53"><a id="__codelineno-0-53" name="__codelineno-0-53"></a><span class="sd">        local_states : dict[str, LocalStateRef]</span>
</span><span id="__span-0-54"><a id="__codelineno-0-54" name="__codelineno-0-54"></a><span class="sd">            Local states.</span>
</span><span id="__span-0-55"><a id="__codelineno-0-55" name="__codelineno-0-55"></a><span class="sd">        round_idx : int</span>
</span><span id="__span-0-56"><a id="__codelineno-0-56" name="__codelineno-0-56"></a><span class="sd">            Round index.</span>
</span><span id="__span-0-57"><a id="__codelineno-0-57" name="__codelineno-0-57"></a><span class="sd">        clean_models : bool</span>
</span><span id="__span-0-58"><a id="__codelineno-0-58" name="__codelineno-0-58"></a><span class="sd">            Whether to clean the models after the computation. (default: ``True``).</span>
</span><span id="__span-0-59"><a id="__codelineno-0-59" name="__codelineno-0-59"></a><span class="sd">            Note that as intermediate steps are very memory consuming, it is recommended</span>
</span><span id="__span-0-60"><a id="__codelineno-0-60" name="__codelineno-0-60"></a><span class="sd">            to clean the models after each step.</span>
</span><span id="__span-0-61"><a id="__codelineno-0-61" name="__codelineno-0-61"></a><span class="sd">        clean_last_model : bool</span>
</span><span id="__span-0-62"><a id="__codelineno-0-62" name="__codelineno-0-62"></a><span class="sd">            Whether to clean the last model. (default: ``False``).</span>
</span><span id="__span-0-63"><a id="__codelineno-0-63" name="__codelineno-0-63"></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="__span-0-64"><a id="__codelineno-0-64" name="__codelineno-0-64"></a>        <span class="c1">#### Build design matrices ####</span>
</span><span id="__span-0-65"><a id="__codelineno-0-65" name="__codelineno-0-65"></a>
</span><span id="__span-0-66"><a id="__codelineno-0-66" name="__codelineno-0-66"></a>        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Building design matrices...&quot;</span><span class="p">)</span>
</span><span id="__span-0-67"><a id="__codelineno-0-67" name="__codelineno-0-67"></a>
</span><span id="__span-0-68"><a id="__codelineno-0-68" name="__codelineno-0-68"></a>        <span class="n">local_states</span><span class="p">,</span> <span class="n">log_mean_shared_states</span><span class="p">,</span> <span class="n">round_idx</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">build_design_matrix</span><span class="p">(</span>
</span><span id="__span-0-69"><a id="__codelineno-0-69" name="__codelineno-0-69"></a>            <span class="n">train_data_nodes</span><span class="p">,</span>
</span><span id="__span-0-70"><a id="__codelineno-0-70" name="__codelineno-0-70"></a>            <span class="n">aggregation_node</span><span class="p">,</span>
</span><span id="__span-0-71"><a id="__codelineno-0-71" name="__codelineno-0-71"></a>            <span class="n">local_states</span><span class="p">,</span>
</span><span id="__span-0-72"><a id="__codelineno-0-72" name="__codelineno-0-72"></a>            <span class="n">round_idx</span><span class="p">,</span>
</span><span id="__span-0-73"><a id="__codelineno-0-73" name="__codelineno-0-73"></a>            <span class="n">clean_models</span><span class="o">=</span><span class="n">clean_models</span><span class="p">,</span>
</span><span id="__span-0-74"><a id="__codelineno-0-74" name="__codelineno-0-74"></a>        <span class="p">)</span>
</span><span id="__span-0-75"><a id="__codelineno-0-75" name="__codelineno-0-75"></a>
</span><span id="__span-0-76"><a id="__codelineno-0-76" name="__codelineno-0-76"></a>        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Finished building design matrices.&quot;</span><span class="p">)</span>
</span><span id="__span-0-77"><a id="__codelineno-0-77" name="__codelineno-0-77"></a>
</span><span id="__span-0-78"><a id="__codelineno-0-78" name="__codelineno-0-78"></a>        <span class="c1">#### Compute size factors ####</span>
</span><span id="__span-0-79"><a id="__codelineno-0-79" name="__codelineno-0-79"></a>        <span class="c1"># Note: in refit mode, this doesn&#39;t recompute size factors,</span>
</span><span id="__span-0-80"><a id="__codelineno-0-80" name="__codelineno-0-80"></a>        <span class="c1"># just the log features</span>
</span><span id="__span-0-81"><a id="__codelineno-0-81" name="__codelineno-0-81"></a>
</span><span id="__span-0-82"><a id="__codelineno-0-82" name="__codelineno-0-82"></a>        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Computing size factors...&quot;</span><span class="p">)</span>
</span><span id="__span-0-83"><a id="__codelineno-0-83" name="__codelineno-0-83"></a>
</span><span id="__span-0-84"><a id="__codelineno-0-84" name="__codelineno-0-84"></a>        <span class="p">(</span>
</span><span id="__span-0-85"><a id="__codelineno-0-85" name="__codelineno-0-85"></a>            <span class="n">local_states</span><span class="p">,</span>
</span><span id="__span-0-86"><a id="__codelineno-0-86" name="__codelineno-0-86"></a>            <span class="n">gram_features_shared_states</span><span class="p">,</span>
</span><span id="__span-0-87"><a id="__codelineno-0-87" name="__codelineno-0-87"></a>            <span class="n">round_idx</span><span class="p">,</span>
</span><span id="__span-0-88"><a id="__codelineno-0-88" name="__codelineno-0-88"></a>        <span class="p">)</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">compute_size_factors</span><span class="p">(</span>
</span><span id="__span-0-89"><a id="__codelineno-0-89" name="__codelineno-0-89"></a>            <span class="n">train_data_nodes</span><span class="p">,</span>
</span><span id="__span-0-90"><a id="__codelineno-0-90" name="__codelineno-0-90"></a>            <span class="n">aggregation_node</span><span class="p">,</span>
</span><span id="__span-0-91"><a id="__codelineno-0-91" name="__codelineno-0-91"></a>            <span class="n">local_states</span><span class="p">,</span>
</span><span id="__span-0-92"><a id="__codelineno-0-92" name="__codelineno-0-92"></a>            <span class="n">shared_states</span><span class="o">=</span><span class="n">log_mean_shared_states</span><span class="p">,</span>
</span><span id="__span-0-93"><a id="__codelineno-0-93" name="__codelineno-0-93"></a>            <span class="n">round_idx</span><span class="o">=</span><span class="n">round_idx</span><span class="p">,</span>
</span><span id="__span-0-94"><a id="__codelineno-0-94" name="__codelineno-0-94"></a>            <span class="n">clean_models</span><span class="o">=</span><span class="n">clean_models</span><span class="p">,</span>
</span><span id="__span-0-95"><a id="__codelineno-0-95" name="__codelineno-0-95"></a>        <span class="p">)</span>
</span><span id="__span-0-96"><a id="__codelineno-0-96" name="__codelineno-0-96"></a>
</span><span id="__span-0-97"><a id="__codelineno-0-97" name="__codelineno-0-97"></a>        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Finished computing size factors.&quot;</span><span class="p">)</span>
</span><span id="__span-0-98"><a id="__codelineno-0-98" name="__codelineno-0-98"></a>
</span><span id="__span-0-99"><a id="__codelineno-0-99" name="__codelineno-0-99"></a>        <span class="c1">#### Compute LFC and dispersions ####</span>
</span><span id="__span-0-100"><a id="__codelineno-0-100" name="__codelineno-0-100"></a>
</span><span id="__span-0-101"><a id="__codelineno-0-101" name="__codelineno-0-101"></a>        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Running LFC and dispersions.&quot;</span><span class="p">)</span>
</span><span id="__span-0-102"><a id="__codelineno-0-102" name="__codelineno-0-102"></a>
</span><span id="__span-0-103"><a id="__codelineno-0-103" name="__codelineno-0-103"></a>        <span class="n">local_states</span><span class="p">,</span> <span class="n">round_idx</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">run_deseq2_lfc_dispersions</span><span class="p">(</span>
</span><span id="__span-0-104"><a id="__codelineno-0-104" name="__codelineno-0-104"></a>            <span class="n">train_data_nodes</span><span class="o">=</span><span class="n">train_data_nodes</span><span class="p">,</span>
</span><span id="__span-0-105"><a id="__codelineno-0-105" name="__codelineno-0-105"></a>            <span class="n">aggregation_node</span><span class="o">=</span><span class="n">aggregation_node</span><span class="p">,</span>
</span><span id="__span-0-106"><a id="__codelineno-0-106" name="__codelineno-0-106"></a>            <span class="n">local_states</span><span class="o">=</span><span class="n">local_states</span><span class="p">,</span>
</span><span id="__span-0-107"><a id="__codelineno-0-107" name="__codelineno-0-107"></a>            <span class="n">gram_features_shared_states</span><span class="o">=</span><span class="n">gram_features_shared_states</span><span class="p">,</span>
</span><span id="__span-0-108"><a id="__codelineno-0-108" name="__codelineno-0-108"></a>            <span class="n">round_idx</span><span class="o">=</span><span class="n">round_idx</span><span class="p">,</span>
</span><span id="__span-0-109"><a id="__codelineno-0-109" name="__codelineno-0-109"></a>            <span class="n">clean_models</span><span class="o">=</span><span class="n">clean_models</span><span class="p">,</span>
</span><span id="__span-0-110"><a id="__codelineno-0-110" name="__codelineno-0-110"></a>        <span class="p">)</span>
</span><span id="__span-0-111"><a id="__codelineno-0-111" name="__codelineno-0-111"></a>
</span><span id="__span-0-112"><a id="__codelineno-0-112" name="__codelineno-0-112"></a>        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Finished running LFC and dispersions.&quot;</span><span class="p">)</span>
</span><span id="__span-0-113"><a id="__codelineno-0-113" name="__codelineno-0-113"></a>
</span><span id="__span-0-114"><a id="__codelineno-0-114" name="__codelineno-0-114"></a>        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Computing Cook distances...&quot;</span><span class="p">)</span>
</span><span id="__span-0-115"><a id="__codelineno-0-115" name="__codelineno-0-115"></a>
</span><span id="__span-0-116"><a id="__codelineno-0-116" name="__codelineno-0-116"></a>        <span class="p">(</span>
</span><span id="__span-0-117"><a id="__codelineno-0-117" name="__codelineno-0-117"></a>            <span class="n">local_states</span><span class="p">,</span>
</span><span id="__span-0-118"><a id="__codelineno-0-118" name="__codelineno-0-118"></a>            <span class="n">cooks_shared_state</span><span class="p">,</span>
</span><span id="__span-0-119"><a id="__codelineno-0-119" name="__codelineno-0-119"></a>            <span class="n">round_idx</span><span class="p">,</span>
</span><span id="__span-0-120"><a id="__codelineno-0-120" name="__codelineno-0-120"></a>        <span class="p">)</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">compute_cook_distance</span><span class="p">(</span>
</span><span id="__span-0-121"><a id="__codelineno-0-121" name="__codelineno-0-121"></a>            <span class="n">train_data_nodes</span><span class="p">,</span>
</span><span id="__span-0-122"><a id="__codelineno-0-122" name="__codelineno-0-122"></a>            <span class="n">aggregation_node</span><span class="p">,</span>
</span><span id="__span-0-123"><a id="__codelineno-0-123" name="__codelineno-0-123"></a>            <span class="n">local_states</span><span class="p">,</span>
</span><span id="__span-0-124"><a id="__codelineno-0-124" name="__codelineno-0-124"></a>            <span class="n">round_idx</span><span class="p">,</span>
</span><span id="__span-0-125"><a id="__codelineno-0-125" name="__codelineno-0-125"></a>            <span class="n">clean_models</span><span class="o">=</span><span class="n">clean_models</span><span class="p">,</span>
</span><span id="__span-0-126"><a id="__codelineno-0-126" name="__codelineno-0-126"></a>        <span class="p">)</span>
</span><span id="__span-0-127"><a id="__codelineno-0-127" name="__codelineno-0-127"></a>
</span><span id="__span-0-128"><a id="__codelineno-0-128" name="__codelineno-0-128"></a>        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Finished computing Cook distances.&quot;</span><span class="p">)</span>
</span><span id="__span-0-129"><a id="__codelineno-0-129" name="__codelineno-0-129"></a>
</span><span id="__span-0-130"><a id="__codelineno-0-130" name="__codelineno-0-130"></a>        <span class="c1">#### Refit cooks if necessary ####</span>
</span><span id="__span-0-131"><a id="__codelineno-0-131" name="__codelineno-0-131"></a>        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">refit_cooks</span><span class="p">:</span>
</span><span id="__span-0-132"><a id="__codelineno-0-132" name="__codelineno-0-132"></a>            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Refitting Cook outliers...&quot;</span><span class="p">)</span>
</span><span id="__span-0-133"><a id="__codelineno-0-133" name="__codelineno-0-133"></a>            <span class="p">(</span>
</span><span id="__span-0-134"><a id="__codelineno-0-134" name="__codelineno-0-134"></a>                <span class="n">local_states</span><span class="p">,</span>
</span><span id="__span-0-135"><a id="__codelineno-0-135" name="__codelineno-0-135"></a>                <span class="n">gram_features_shared_states</span><span class="p">,</span>
</span><span id="__span-0-136"><a id="__codelineno-0-136" name="__codelineno-0-136"></a>                <span class="n">round_idx</span><span class="p">,</span>
</span><span id="__span-0-137"><a id="__codelineno-0-137" name="__codelineno-0-137"></a>            <span class="p">)</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">replace_outliers</span><span class="p">(</span>
</span><span id="__span-0-138"><a id="__codelineno-0-138" name="__codelineno-0-138"></a>                <span class="n">train_data_nodes</span><span class="p">,</span>
</span><span id="__span-0-139"><a id="__codelineno-0-139" name="__codelineno-0-139"></a>                <span class="n">aggregation_node</span><span class="p">,</span>
</span><span id="__span-0-140"><a id="__codelineno-0-140" name="__codelineno-0-140"></a>                <span class="n">local_states</span><span class="p">,</span>
</span><span id="__span-0-141"><a id="__codelineno-0-141" name="__codelineno-0-141"></a>                <span class="n">cooks_shared_state</span><span class="p">,</span>
</span><span id="__span-0-142"><a id="__codelineno-0-142" name="__codelineno-0-142"></a>                <span class="n">round_idx</span><span class="p">,</span>
</span><span id="__span-0-143"><a id="__codelineno-0-143" name="__codelineno-0-143"></a>                <span class="n">clean_models</span><span class="o">=</span><span class="n">clean_models</span><span class="p">,</span>
</span><span id="__span-0-144"><a id="__codelineno-0-144" name="__codelineno-0-144"></a>            <span class="p">)</span>
</span><span id="__span-0-145"><a id="__codelineno-0-145" name="__codelineno-0-145"></a>
</span><span id="__span-0-146"><a id="__codelineno-0-146" name="__codelineno-0-146"></a>            <span class="n">local_states</span><span class="p">,</span> <span class="n">round_idx</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">run_deseq2_lfc_dispersions</span><span class="p">(</span>
</span><span id="__span-0-147"><a id="__codelineno-0-147" name="__codelineno-0-147"></a>                <span class="n">train_data_nodes</span><span class="o">=</span><span class="n">train_data_nodes</span><span class="p">,</span>
</span><span id="__span-0-148"><a id="__codelineno-0-148" name="__codelineno-0-148"></a>                <span class="n">aggregation_node</span><span class="o">=</span><span class="n">aggregation_node</span><span class="p">,</span>
</span><span id="__span-0-149"><a id="__codelineno-0-149" name="__codelineno-0-149"></a>                <span class="n">local_states</span><span class="o">=</span><span class="n">local_states</span><span class="p">,</span>
</span><span id="__span-0-150"><a id="__codelineno-0-150" name="__codelineno-0-150"></a>                <span class="n">gram_features_shared_states</span><span class="o">=</span><span class="n">gram_features_shared_states</span><span class="p">,</span>
</span><span id="__span-0-151"><a id="__codelineno-0-151" name="__codelineno-0-151"></a>                <span class="n">round_idx</span><span class="o">=</span><span class="n">round_idx</span><span class="p">,</span>
</span><span id="__span-0-152"><a id="__codelineno-0-152" name="__codelineno-0-152"></a>                <span class="n">clean_models</span><span class="o">=</span><span class="n">clean_models</span><span class="p">,</span>
</span><span id="__span-0-153"><a id="__codelineno-0-153" name="__codelineno-0-153"></a>                <span class="n">refit_mode</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
</span><span id="__span-0-154"><a id="__codelineno-0-154" name="__codelineno-0-154"></a>            <span class="p">)</span>
</span><span id="__span-0-155"><a id="__codelineno-0-155" name="__codelineno-0-155"></a>            <span class="c1"># Replace values in the main ``local_adata`` object</span>
</span><span id="__span-0-156"><a id="__codelineno-0-156" name="__codelineno-0-156"></a>            <span class="n">local_states</span><span class="p">,</span> <span class="n">round_idx</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">replace_refitted_values</span><span class="p">(</span>
</span><span id="__span-0-157"><a id="__codelineno-0-157" name="__codelineno-0-157"></a>                <span class="n">train_data_nodes</span><span class="o">=</span><span class="n">train_data_nodes</span><span class="p">,</span>
</span><span id="__span-0-158"><a id="__codelineno-0-158" name="__codelineno-0-158"></a>                <span class="n">aggregation_node</span><span class="o">=</span><span class="n">aggregation_node</span><span class="p">,</span>
</span><span id="__span-0-159"><a id="__codelineno-0-159" name="__codelineno-0-159"></a>                <span class="n">local_states</span><span class="o">=</span><span class="n">local_states</span><span class="p">,</span>
</span><span id="__span-0-160"><a id="__codelineno-0-160" name="__codelineno-0-160"></a>                <span class="n">round_idx</span><span class="o">=</span><span class="n">round_idx</span><span class="p">,</span>
</span><span id="__span-0-161"><a id="__codelineno-0-161" name="__codelineno-0-161"></a>                <span class="n">clean_models</span><span class="o">=</span><span class="n">clean_models</span><span class="p">,</span>
</span><span id="__span-0-162"><a id="__codelineno-0-162" name="__codelineno-0-162"></a>            <span class="p">)</span>
</span><span id="__span-0-163"><a id="__codelineno-0-163" name="__codelineno-0-163"></a>
</span><span id="__span-0-164"><a id="__codelineno-0-164" name="__codelineno-0-164"></a>            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Finished refitting Cook outliers.&quot;</span><span class="p">)</span>
</span><span id="__span-0-165"><a id="__codelineno-0-165" name="__codelineno-0-165"></a>
</span><span id="__span-0-166"><a id="__codelineno-0-166" name="__codelineno-0-166"></a>        <span class="c1">#### Compute DESeq2 statistics ####</span>
</span><span id="__span-0-167"><a id="__codelineno-0-167" name="__codelineno-0-167"></a>
</span><span id="__span-0-168"><a id="__codelineno-0-168" name="__codelineno-0-168"></a>        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Running DESeq2 statistics.&quot;</span><span class="p">)</span>
</span><span id="__span-0-169"><a id="__codelineno-0-169" name="__codelineno-0-169"></a>
</span><span id="__span-0-170"><a id="__codelineno-0-170" name="__codelineno-0-170"></a>        <span class="n">local_states</span><span class="p">,</span> <span class="n">round_idx</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">run_deseq2_stats</span><span class="p">(</span>
</span><span id="__span-0-171"><a id="__codelineno-0-171" name="__codelineno-0-171"></a>            <span class="n">train_data_nodes</span><span class="p">,</span>
</span><span id="__span-0-172"><a id="__codelineno-0-172" name="__codelineno-0-172"></a>            <span class="n">aggregation_node</span><span class="p">,</span>
</span><span id="__span-0-173"><a id="__codelineno-0-173" name="__codelineno-0-173"></a>            <span class="n">local_states</span><span class="p">,</span>
</span><span id="__span-0-174"><a id="__codelineno-0-174" name="__codelineno-0-174"></a>            <span class="n">round_idx</span><span class="p">,</span>
</span><span id="__span-0-175"><a id="__codelineno-0-175" name="__codelineno-0-175"></a>            <span class="n">clean_models</span><span class="o">=</span><span class="n">clean_models</span><span class="p">,</span>
</span><span id="__span-0-176"><a id="__codelineno-0-176" name="__codelineno-0-176"></a>        <span class="p">)</span>
</span><span id="__span-0-177"><a id="__codelineno-0-177" name="__codelineno-0-177"></a>
</span><span id="__span-0-178"><a id="__codelineno-0-178" name="__codelineno-0-178"></a>        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Finished running DESeq2 statistics.&quot;</span><span class="p">)</span>
</span><span id="__span-0-179"><a id="__codelineno-0-179" name="__codelineno-0-179"></a>
</span><span id="__span-0-180"><a id="__codelineno-0-180" name="__codelineno-0-180"></a>        <span class="c1"># Build the results that will be downloaded at the end of the pipeline.</span>
</span><span id="__span-0-181"><a id="__codelineno-0-181" name="__codelineno-0-181"></a>
</span><span id="__span-0-182"><a id="__codelineno-0-182" name="__codelineno-0-182"></a>        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Saving pipeline results.&quot;</span><span class="p">)</span>
</span><span id="__span-0-183"><a id="__codelineno-0-183" name="__codelineno-0-183"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">save_pipeline_results</span><span class="p">(</span>
</span><span id="__span-0-184"><a id="__codelineno-0-184" name="__codelineno-0-184"></a>            <span class="n">train_data_nodes</span><span class="p">,</span>
</span><span id="__span-0-185"><a id="__codelineno-0-185" name="__codelineno-0-185"></a>            <span class="n">aggregation_node</span><span class="p">,</span>
</span><span id="__span-0-186"><a id="__codelineno-0-186" name="__codelineno-0-186"></a>            <span class="n">local_states</span><span class="p">,</span>
</span><span id="__span-0-187"><a id="__codelineno-0-187" name="__codelineno-0-187"></a>            <span class="n">round_idx</span><span class="p">,</span>
</span><span id="__span-0-188"><a id="__codelineno-0-188" name="__codelineno-0-188"></a>            <span class="n">clean_models</span><span class="o">=</span><span class="n">clean_models</span><span class="p">,</span>
</span><span id="__span-0-189"><a id="__codelineno-0-189" name="__codelineno-0-189"></a>        <span class="p">)</span>
</span><span id="__span-0-190"><a id="__codelineno-0-190" name="__codelineno-0-190"></a>
</span><span id="__span-0-191"><a id="__codelineno-0-191" name="__codelineno-0-191"></a>        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Finished saving pipeline results.&quot;</span><span class="p">)</span>
</span></code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="fedpydeseq2.core.deseq2_core.deseq2_full_pipe.DESeq2FullPipe.run_deseq_pipe" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">run_deseq_pipe</span><span class="p">(</span><span class="n">train_data_nodes</span><span class="p">,</span> <span class="n">aggregation_node</span><span class="p">,</span> <span class="n">local_states</span><span class="p">,</span> <span class="n">round_idx</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">clean_models</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">clean_last_model</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Run the DESeq2 pipeline.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>train_data_nodes</code>
            </td>
            <td>
                  <code><span title="list">list</span>[<span title="substrafl.nodes.TrainDataNode">TrainDataNode</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List of the train nodes.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>aggregation_node</code>
            </td>
            <td>
                  <code><span title="substrafl.nodes.AggregationNode">AggregationNode</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Aggregation node.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>local_states</code>
            </td>
            <td>
                  <code><span title="dict">dict</span>[<span title="str">str</span>, <span title="substrafl.nodes.references.local_state.LocalStateRef">LocalStateRef</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Local states.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>round_idx</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Round index.</p>
              </div>
            </td>
            <td>
                  <code>0</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>clean_models</code>
            </td>
            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether to clean the models after the computation. (default: <code>True</code>).
Note that as intermediate steps are very memory consuming, it is recommended
to clean the models after each step.</p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>clean_last_model</code>
            </td>
            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether to clean the last model. (default: <code>False</code>).</p>
              </div>
            </td>
            <td>
                  <code>False</code>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>fedpydeseq2/core/deseq2_core/deseq2_full_pipe.py</code></summary>
              <div class="language-python highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-35"> 35</a></span>
<span class="normal"><a href="#__codelineno-0-36"> 36</a></span>
<span class="normal"><a href="#__codelineno-0-37"> 37</a></span>
<span class="normal"><a href="#__codelineno-0-38"> 38</a></span>
<span class="normal"><a href="#__codelineno-0-39"> 39</a></span>
<span class="normal"><a href="#__codelineno-0-40"> 40</a></span>
<span class="normal"><a href="#__codelineno-0-41"> 41</a></span>
<span class="normal"><a href="#__codelineno-0-42"> 42</a></span>
<span class="normal"><a href="#__codelineno-0-43"> 43</a></span>
<span class="normal"><a href="#__codelineno-0-44"> 44</a></span>
<span class="normal"><a href="#__codelineno-0-45"> 45</a></span>
<span class="normal"><a href="#__codelineno-0-46"> 46</a></span>
<span class="normal"><a href="#__codelineno-0-47"> 47</a></span>
<span class="normal"><a href="#__codelineno-0-48"> 48</a></span>
<span class="normal"><a href="#__codelineno-0-49"> 49</a></span>
<span class="normal"><a href="#__codelineno-0-50"> 50</a></span>
<span class="normal"><a href="#__codelineno-0-51"> 51</a></span>
<span class="normal"><a href="#__codelineno-0-52"> 52</a></span>
<span class="normal"><a href="#__codelineno-0-53"> 53</a></span>
<span class="normal"><a href="#__codelineno-0-54"> 54</a></span>
<span class="normal"><a href="#__codelineno-0-55"> 55</a></span>
<span class="normal"><a href="#__codelineno-0-56"> 56</a></span>
<span class="normal"><a href="#__codelineno-0-57"> 57</a></span>
<span class="normal"><a href="#__codelineno-0-58"> 58</a></span>
<span class="normal"><a href="#__codelineno-0-59"> 59</a></span>
<span class="normal"><a href="#__codelineno-0-60"> 60</a></span>
<span class="normal"><a href="#__codelineno-0-61"> 61</a></span>
<span class="normal"><a href="#__codelineno-0-62"> 62</a></span>
<span class="normal"><a href="#__codelineno-0-63"> 63</a></span>
<span class="normal"><a href="#__codelineno-0-64"> 64</a></span>
<span class="normal"><a href="#__codelineno-0-65"> 65</a></span>
<span class="normal"><a href="#__codelineno-0-66"> 66</a></span>
<span class="normal"><a href="#__codelineno-0-67"> 67</a></span>
<span class="normal"><a href="#__codelineno-0-68"> 68</a></span>
<span class="normal"><a href="#__codelineno-0-69"> 69</a></span>
<span class="normal"><a href="#__codelineno-0-70"> 70</a></span>
<span class="normal"><a href="#__codelineno-0-71"> 71</a></span>
<span class="normal"><a href="#__codelineno-0-72"> 72</a></span>
<span class="normal"><a href="#__codelineno-0-73"> 73</a></span>
<span class="normal"><a href="#__codelineno-0-74"> 74</a></span>
<span class="normal"><a href="#__codelineno-0-75"> 75</a></span>
<span class="normal"><a href="#__codelineno-0-76"> 76</a></span>
<span class="normal"><a href="#__codelineno-0-77"> 77</a></span>
<span class="normal"><a href="#__codelineno-0-78"> 78</a></span>
<span class="normal"><a href="#__codelineno-0-79"> 79</a></span>
<span class="normal"><a href="#__codelineno-0-80"> 80</a></span>
<span class="normal"><a href="#__codelineno-0-81"> 81</a></span>
<span class="normal"><a href="#__codelineno-0-82"> 82</a></span>
<span class="normal"><a href="#__codelineno-0-83"> 83</a></span>
<span class="normal"><a href="#__codelineno-0-84"> 84</a></span>
<span class="normal"><a href="#__codelineno-0-85"> 85</a></span>
<span class="normal"><a href="#__codelineno-0-86"> 86</a></span>
<span class="normal"><a href="#__codelineno-0-87"> 87</a></span>
<span class="normal"><a href="#__codelineno-0-88"> 88</a></span>
<span class="normal"><a href="#__codelineno-0-89"> 89</a></span>
<span class="normal"><a href="#__codelineno-0-90"> 90</a></span>
<span class="normal"><a href="#__codelineno-0-91"> 91</a></span>
<span class="normal"><a href="#__codelineno-0-92"> 92</a></span>
<span class="normal"><a href="#__codelineno-0-93"> 93</a></span>
<span class="normal"><a href="#__codelineno-0-94"> 94</a></span>
<span class="normal"><a href="#__codelineno-0-95"> 95</a></span>
<span class="normal"><a href="#__codelineno-0-96"> 96</a></span>
<span class="normal"><a href="#__codelineno-0-97"> 97</a></span>
<span class="normal"><a href="#__codelineno-0-98"> 98</a></span>
<span class="normal"><a href="#__codelineno-0-99"> 99</a></span>
<span class="normal"><a href="#__codelineno-0-100">100</a></span>
<span class="normal"><a href="#__codelineno-0-101">101</a></span>
<span class="normal"><a href="#__codelineno-0-102">102</a></span>
<span class="normal"><a href="#__codelineno-0-103">103</a></span>
<span class="normal"><a href="#__codelineno-0-104">104</a></span>
<span class="normal"><a href="#__codelineno-0-105">105</a></span>
<span class="normal"><a href="#__codelineno-0-106">106</a></span>
<span class="normal"><a href="#__codelineno-0-107">107</a></span>
<span class="normal"><a href="#__codelineno-0-108">108</a></span>
<span class="normal"><a href="#__codelineno-0-109">109</a></span>
<span class="normal"><a href="#__codelineno-0-110">110</a></span>
<span class="normal"><a href="#__codelineno-0-111">111</a></span>
<span class="normal"><a href="#__codelineno-0-112">112</a></span>
<span class="normal"><a href="#__codelineno-0-113">113</a></span>
<span class="normal"><a href="#__codelineno-0-114">114</a></span>
<span class="normal"><a href="#__codelineno-0-115">115</a></span>
<span class="normal"><a href="#__codelineno-0-116">116</a></span>
<span class="normal"><a href="#__codelineno-0-117">117</a></span>
<span class="normal"><a href="#__codelineno-0-118">118</a></span>
<span class="normal"><a href="#__codelineno-0-119">119</a></span>
<span class="normal"><a href="#__codelineno-0-120">120</a></span>
<span class="normal"><a href="#__codelineno-0-121">121</a></span>
<span class="normal"><a href="#__codelineno-0-122">122</a></span>
<span class="normal"><a href="#__codelineno-0-123">123</a></span>
<span class="normal"><a href="#__codelineno-0-124">124</a></span>
<span class="normal"><a href="#__codelineno-0-125">125</a></span>
<span class="normal"><a href="#__codelineno-0-126">126</a></span>
<span class="normal"><a href="#__codelineno-0-127">127</a></span>
<span class="normal"><a href="#__codelineno-0-128">128</a></span>
<span class="normal"><a href="#__codelineno-0-129">129</a></span>
<span class="normal"><a href="#__codelineno-0-130">130</a></span>
<span class="normal"><a href="#__codelineno-0-131">131</a></span>
<span class="normal"><a href="#__codelineno-0-132">132</a></span>
<span class="normal"><a href="#__codelineno-0-133">133</a></span>
<span class="normal"><a href="#__codelineno-0-134">134</a></span>
<span class="normal"><a href="#__codelineno-0-135">135</a></span>
<span class="normal"><a href="#__codelineno-0-136">136</a></span>
<span class="normal"><a href="#__codelineno-0-137">137</a></span>
<span class="normal"><a href="#__codelineno-0-138">138</a></span>
<span class="normal"><a href="#__codelineno-0-139">139</a></span>
<span class="normal"><a href="#__codelineno-0-140">140</a></span>
<span class="normal"><a href="#__codelineno-0-141">141</a></span>
<span class="normal"><a href="#__codelineno-0-142">142</a></span>
<span class="normal"><a href="#__codelineno-0-143">143</a></span>
<span class="normal"><a href="#__codelineno-0-144">144</a></span>
<span class="normal"><a href="#__codelineno-0-145">145</a></span>
<span class="normal"><a href="#__codelineno-0-146">146</a></span>
<span class="normal"><a href="#__codelineno-0-147">147</a></span>
<span class="normal"><a href="#__codelineno-0-148">148</a></span>
<span class="normal"><a href="#__codelineno-0-149">149</a></span>
<span class="normal"><a href="#__codelineno-0-150">150</a></span>
<span class="normal"><a href="#__codelineno-0-151">151</a></span>
<span class="normal"><a href="#__codelineno-0-152">152</a></span>
<span class="normal"><a href="#__codelineno-0-153">153</a></span>
<span class="normal"><a href="#__codelineno-0-154">154</a></span>
<span class="normal"><a href="#__codelineno-0-155">155</a></span>
<span class="normal"><a href="#__codelineno-0-156">156</a></span>
<span class="normal"><a href="#__codelineno-0-157">157</a></span>
<span class="normal"><a href="#__codelineno-0-158">158</a></span>
<span class="normal"><a href="#__codelineno-0-159">159</a></span>
<span class="normal"><a href="#__codelineno-0-160">160</a></span>
<span class="normal"><a href="#__codelineno-0-161">161</a></span>
<span class="normal"><a href="#__codelineno-0-162">162</a></span>
<span class="normal"><a href="#__codelineno-0-163">163</a></span>
<span class="normal"><a href="#__codelineno-0-164">164</a></span>
<span class="normal"><a href="#__codelineno-0-165">165</a></span>
<span class="normal"><a href="#__codelineno-0-166">166</a></span>
<span class="normal"><a href="#__codelineno-0-167">167</a></span>
<span class="normal"><a href="#__codelineno-0-168">168</a></span>
<span class="normal"><a href="#__codelineno-0-169">169</a></span>
<span class="normal"><a href="#__codelineno-0-170">170</a></span>
<span class="normal"><a href="#__codelineno-0-171">171</a></span>
<span class="normal"><a href="#__codelineno-0-172">172</a></span>
<span class="normal"><a href="#__codelineno-0-173">173</a></span>
<span class="normal"><a href="#__codelineno-0-174">174</a></span>
<span class="normal"><a href="#__codelineno-0-175">175</a></span>
<span class="normal"><a href="#__codelineno-0-176">176</a></span>
<span class="normal"><a href="#__codelineno-0-177">177</a></span>
<span class="normal"><a href="#__codelineno-0-178">178</a></span>
<span class="normal"><a href="#__codelineno-0-179">179</a></span>
<span class="normal"><a href="#__codelineno-0-180">180</a></span>
<span class="normal"><a href="#__codelineno-0-181">181</a></span>
<span class="normal"><a href="#__codelineno-0-182">182</a></span>
<span class="normal"><a href="#__codelineno-0-183">183</a></span>
<span class="normal"><a href="#__codelineno-0-184">184</a></span>
<span class="normal"><a href="#__codelineno-0-185">185</a></span>
<span class="normal"><a href="#__codelineno-0-186">186</a></span>
<span class="normal"><a href="#__codelineno-0-187">187</a></span>
<span class="normal"><a href="#__codelineno-0-188">188</a></span>
<span class="normal"><a href="#__codelineno-0-189">189</a></span>
<span class="normal"><a href="#__codelineno-0-190">190</a></span>
<span class="normal"><a href="#__codelineno-0-191">191</a></span></pre></div></td><td class="code"><div><pre><span></span><code><span id="__span-0-35"><a id="__codelineno-0-35" name="__codelineno-0-35"></a><span class="nd">@log_organisation_method</span>
</span><span id="__span-0-36"><a id="__codelineno-0-36" name="__codelineno-0-36"></a><span class="k">def</span><span class="w"> </span><span class="nf">run_deseq_pipe</span><span class="p">(</span>
</span><span id="__span-0-37"><a id="__codelineno-0-37" name="__codelineno-0-37"></a>    <span class="bp">self</span><span class="p">,</span>
</span><span id="__span-0-38"><a id="__codelineno-0-38" name="__codelineno-0-38"></a>    <span class="n">train_data_nodes</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="n">TrainDataNode</span><span class="p">],</span>
</span><span id="__span-0-39"><a id="__codelineno-0-39" name="__codelineno-0-39"></a>    <span class="n">aggregation_node</span><span class="p">:</span> <span class="n">AggregationNode</span><span class="p">,</span>
</span><span id="__span-0-40"><a id="__codelineno-0-40" name="__codelineno-0-40"></a>    <span class="n">local_states</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">LocalStateRef</span><span class="p">],</span>
</span><span id="__span-0-41"><a id="__codelineno-0-41" name="__codelineno-0-41"></a>    <span class="n">round_idx</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>
</span><span id="__span-0-42"><a id="__codelineno-0-42" name="__codelineno-0-42"></a>    <span class="n">clean_models</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
</span><span id="__span-0-43"><a id="__codelineno-0-43" name="__codelineno-0-43"></a>    <span class="n">clean_last_model</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
</span><span id="__span-0-44"><a id="__codelineno-0-44" name="__codelineno-0-44"></a><span class="p">):</span>
</span><span id="__span-0-45"><a id="__codelineno-0-45" name="__codelineno-0-45"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;Run the DESeq2 pipeline.</span>
</span><span id="__span-0-46"><a id="__codelineno-0-46" name="__codelineno-0-46"></a>
</span><span id="__span-0-47"><a id="__codelineno-0-47" name="__codelineno-0-47"></a><span class="sd">    Parameters</span>
</span><span id="__span-0-48"><a id="__codelineno-0-48" name="__codelineno-0-48"></a><span class="sd">    ----------</span>
</span><span id="__span-0-49"><a id="__codelineno-0-49" name="__codelineno-0-49"></a><span class="sd">    train_data_nodes : list[TrainDataNode]</span>
</span><span id="__span-0-50"><a id="__codelineno-0-50" name="__codelineno-0-50"></a><span class="sd">        List of the train nodes.</span>
</span><span id="__span-0-51"><a id="__codelineno-0-51" name="__codelineno-0-51"></a><span class="sd">    aggregation_node : AggregationNode</span>
</span><span id="__span-0-52"><a id="__codelineno-0-52" name="__codelineno-0-52"></a><span class="sd">        Aggregation node.</span>
</span><span id="__span-0-53"><a id="__codelineno-0-53" name="__codelineno-0-53"></a><span class="sd">    local_states : dict[str, LocalStateRef]</span>
</span><span id="__span-0-54"><a id="__codelineno-0-54" name="__codelineno-0-54"></a><span class="sd">        Local states.</span>
</span><span id="__span-0-55"><a id="__codelineno-0-55" name="__codelineno-0-55"></a><span class="sd">    round_idx : int</span>
</span><span id="__span-0-56"><a id="__codelineno-0-56" name="__codelineno-0-56"></a><span class="sd">        Round index.</span>
</span><span id="__span-0-57"><a id="__codelineno-0-57" name="__codelineno-0-57"></a><span class="sd">    clean_models : bool</span>
</span><span id="__span-0-58"><a id="__codelineno-0-58" name="__codelineno-0-58"></a><span class="sd">        Whether to clean the models after the computation. (default: ``True``).</span>
</span><span id="__span-0-59"><a id="__codelineno-0-59" name="__codelineno-0-59"></a><span class="sd">        Note that as intermediate steps are very memory consuming, it is recommended</span>
</span><span id="__span-0-60"><a id="__codelineno-0-60" name="__codelineno-0-60"></a><span class="sd">        to clean the models after each step.</span>
</span><span id="__span-0-61"><a id="__codelineno-0-61" name="__codelineno-0-61"></a><span class="sd">    clean_last_model : bool</span>
</span><span id="__span-0-62"><a id="__codelineno-0-62" name="__codelineno-0-62"></a><span class="sd">        Whether to clean the last model. (default: ``False``).</span>
</span><span id="__span-0-63"><a id="__codelineno-0-63" name="__codelineno-0-63"></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="__span-0-64"><a id="__codelineno-0-64" name="__codelineno-0-64"></a>    <span class="c1">#### Build design matrices ####</span>
</span><span id="__span-0-65"><a id="__codelineno-0-65" name="__codelineno-0-65"></a>
</span><span id="__span-0-66"><a id="__codelineno-0-66" name="__codelineno-0-66"></a>    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Building design matrices...&quot;</span><span class="p">)</span>
</span><span id="__span-0-67"><a id="__codelineno-0-67" name="__codelineno-0-67"></a>
</span><span id="__span-0-68"><a id="__codelineno-0-68" name="__codelineno-0-68"></a>    <span class="n">local_states</span><span class="p">,</span> <span class="n">log_mean_shared_states</span><span class="p">,</span> <span class="n">round_idx</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">build_design_matrix</span><span class="p">(</span>
</span><span id="__span-0-69"><a id="__codelineno-0-69" name="__codelineno-0-69"></a>        <span class="n">train_data_nodes</span><span class="p">,</span>
</span><span id="__span-0-70"><a id="__codelineno-0-70" name="__codelineno-0-70"></a>        <span class="n">aggregation_node</span><span class="p">,</span>
</span><span id="__span-0-71"><a id="__codelineno-0-71" name="__codelineno-0-71"></a>        <span class="n">local_states</span><span class="p">,</span>
</span><span id="__span-0-72"><a id="__codelineno-0-72" name="__codelineno-0-72"></a>        <span class="n">round_idx</span><span class="p">,</span>
</span><span id="__span-0-73"><a id="__codelineno-0-73" name="__codelineno-0-73"></a>        <span class="n">clean_models</span><span class="o">=</span><span class="n">clean_models</span><span class="p">,</span>
</span><span id="__span-0-74"><a id="__codelineno-0-74" name="__codelineno-0-74"></a>    <span class="p">)</span>
</span><span id="__span-0-75"><a id="__codelineno-0-75" name="__codelineno-0-75"></a>
</span><span id="__span-0-76"><a id="__codelineno-0-76" name="__codelineno-0-76"></a>    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Finished building design matrices.&quot;</span><span class="p">)</span>
</span><span id="__span-0-77"><a id="__codelineno-0-77" name="__codelineno-0-77"></a>
</span><span id="__span-0-78"><a id="__codelineno-0-78" name="__codelineno-0-78"></a>    <span class="c1">#### Compute size factors ####</span>
</span><span id="__span-0-79"><a id="__codelineno-0-79" name="__codelineno-0-79"></a>    <span class="c1"># Note: in refit mode, this doesn&#39;t recompute size factors,</span>
</span><span id="__span-0-80"><a id="__codelineno-0-80" name="__codelineno-0-80"></a>    <span class="c1"># just the log features</span>
</span><span id="__span-0-81"><a id="__codelineno-0-81" name="__codelineno-0-81"></a>
</span><span id="__span-0-82"><a id="__codelineno-0-82" name="__codelineno-0-82"></a>    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Computing size factors...&quot;</span><span class="p">)</span>
</span><span id="__span-0-83"><a id="__codelineno-0-83" name="__codelineno-0-83"></a>
</span><span id="__span-0-84"><a id="__codelineno-0-84" name="__codelineno-0-84"></a>    <span class="p">(</span>
</span><span id="__span-0-85"><a id="__codelineno-0-85" name="__codelineno-0-85"></a>        <span class="n">local_states</span><span class="p">,</span>
</span><span id="__span-0-86"><a id="__codelineno-0-86" name="__codelineno-0-86"></a>        <span class="n">gram_features_shared_states</span><span class="p">,</span>
</span><span id="__span-0-87"><a id="__codelineno-0-87" name="__codelineno-0-87"></a>        <span class="n">round_idx</span><span class="p">,</span>
</span><span id="__span-0-88"><a id="__codelineno-0-88" name="__codelineno-0-88"></a>    <span class="p">)</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">compute_size_factors</span><span class="p">(</span>
</span><span id="__span-0-89"><a id="__codelineno-0-89" name="__codelineno-0-89"></a>        <span class="n">train_data_nodes</span><span class="p">,</span>
</span><span id="__span-0-90"><a id="__codelineno-0-90" name="__codelineno-0-90"></a>        <span class="n">aggregation_node</span><span class="p">,</span>
</span><span id="__span-0-91"><a id="__codelineno-0-91" name="__codelineno-0-91"></a>        <span class="n">local_states</span><span class="p">,</span>
</span><span id="__span-0-92"><a id="__codelineno-0-92" name="__codelineno-0-92"></a>        <span class="n">shared_states</span><span class="o">=</span><span class="n">log_mean_shared_states</span><span class="p">,</span>
</span><span id="__span-0-93"><a id="__codelineno-0-93" name="__codelineno-0-93"></a>        <span class="n">round_idx</span><span class="o">=</span><span class="n">round_idx</span><span class="p">,</span>
</span><span id="__span-0-94"><a id="__codelineno-0-94" name="__codelineno-0-94"></a>        <span class="n">clean_models</span><span class="o">=</span><span class="n">clean_models</span><span class="p">,</span>
</span><span id="__span-0-95"><a id="__codelineno-0-95" name="__codelineno-0-95"></a>    <span class="p">)</span>
</span><span id="__span-0-96"><a id="__codelineno-0-96" name="__codelineno-0-96"></a>
</span><span id="__span-0-97"><a id="__codelineno-0-97" name="__codelineno-0-97"></a>    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Finished computing size factors.&quot;</span><span class="p">)</span>
</span><span id="__span-0-98"><a id="__codelineno-0-98" name="__codelineno-0-98"></a>
</span><span id="__span-0-99"><a id="__codelineno-0-99" name="__codelineno-0-99"></a>    <span class="c1">#### Compute LFC and dispersions ####</span>
</span><span id="__span-0-100"><a id="__codelineno-0-100" name="__codelineno-0-100"></a>
</span><span id="__span-0-101"><a id="__codelineno-0-101" name="__codelineno-0-101"></a>    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Running LFC and dispersions.&quot;</span><span class="p">)</span>
</span><span id="__span-0-102"><a id="__codelineno-0-102" name="__codelineno-0-102"></a>
</span><span id="__span-0-103"><a id="__codelineno-0-103" name="__codelineno-0-103"></a>    <span class="n">local_states</span><span class="p">,</span> <span class="n">round_idx</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">run_deseq2_lfc_dispersions</span><span class="p">(</span>
</span><span id="__span-0-104"><a id="__codelineno-0-104" name="__codelineno-0-104"></a>        <span class="n">train_data_nodes</span><span class="o">=</span><span class="n">train_data_nodes</span><span class="p">,</span>
</span><span id="__span-0-105"><a id="__codelineno-0-105" name="__codelineno-0-105"></a>        <span class="n">aggregation_node</span><span class="o">=</span><span class="n">aggregation_node</span><span class="p">,</span>
</span><span id="__span-0-106"><a id="__codelineno-0-106" name="__codelineno-0-106"></a>        <span class="n">local_states</span><span class="o">=</span><span class="n">local_states</span><span class="p">,</span>
</span><span id="__span-0-107"><a id="__codelineno-0-107" name="__codelineno-0-107"></a>        <span class="n">gram_features_shared_states</span><span class="o">=</span><span class="n">gram_features_shared_states</span><span class="p">,</span>
</span><span id="__span-0-108"><a id="__codelineno-0-108" name="__codelineno-0-108"></a>        <span class="n">round_idx</span><span class="o">=</span><span class="n">round_idx</span><span class="p">,</span>
</span><span id="__span-0-109"><a id="__codelineno-0-109" name="__codelineno-0-109"></a>        <span class="n">clean_models</span><span class="o">=</span><span class="n">clean_models</span><span class="p">,</span>
</span><span id="__span-0-110"><a id="__codelineno-0-110" name="__codelineno-0-110"></a>    <span class="p">)</span>
</span><span id="__span-0-111"><a id="__codelineno-0-111" name="__codelineno-0-111"></a>
</span><span id="__span-0-112"><a id="__codelineno-0-112" name="__codelineno-0-112"></a>    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Finished running LFC and dispersions.&quot;</span><span class="p">)</span>
</span><span id="__span-0-113"><a id="__codelineno-0-113" name="__codelineno-0-113"></a>
</span><span id="__span-0-114"><a id="__codelineno-0-114" name="__codelineno-0-114"></a>    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Computing Cook distances...&quot;</span><span class="p">)</span>
</span><span id="__span-0-115"><a id="__codelineno-0-115" name="__codelineno-0-115"></a>
</span><span id="__span-0-116"><a id="__codelineno-0-116" name="__codelineno-0-116"></a>    <span class="p">(</span>
</span><span id="__span-0-117"><a id="__codelineno-0-117" name="__codelineno-0-117"></a>        <span class="n">local_states</span><span class="p">,</span>
</span><span id="__span-0-118"><a id="__codelineno-0-118" name="__codelineno-0-118"></a>        <span class="n">cooks_shared_state</span><span class="p">,</span>
</span><span id="__span-0-119"><a id="__codelineno-0-119" name="__codelineno-0-119"></a>        <span class="n">round_idx</span><span class="p">,</span>
</span><span id="__span-0-120"><a id="__codelineno-0-120" name="__codelineno-0-120"></a>    <span class="p">)</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">compute_cook_distance</span><span class="p">(</span>
</span><span id="__span-0-121"><a id="__codelineno-0-121" name="__codelineno-0-121"></a>        <span class="n">train_data_nodes</span><span class="p">,</span>
</span><span id="__span-0-122"><a id="__codelineno-0-122" name="__codelineno-0-122"></a>        <span class="n">aggregation_node</span><span class="p">,</span>
</span><span id="__span-0-123"><a id="__codelineno-0-123" name="__codelineno-0-123"></a>        <span class="n">local_states</span><span class="p">,</span>
</span><span id="__span-0-124"><a id="__codelineno-0-124" name="__codelineno-0-124"></a>        <span class="n">round_idx</span><span class="p">,</span>
</span><span id="__span-0-125"><a id="__codelineno-0-125" name="__codelineno-0-125"></a>        <span class="n">clean_models</span><span class="o">=</span><span class="n">clean_models</span><span class="p">,</span>
</span><span id="__span-0-126"><a id="__codelineno-0-126" name="__codelineno-0-126"></a>    <span class="p">)</span>
</span><span id="__span-0-127"><a id="__codelineno-0-127" name="__codelineno-0-127"></a>
</span><span id="__span-0-128"><a id="__codelineno-0-128" name="__codelineno-0-128"></a>    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Finished computing Cook distances.&quot;</span><span class="p">)</span>
</span><span id="__span-0-129"><a id="__codelineno-0-129" name="__codelineno-0-129"></a>
</span><span id="__span-0-130"><a id="__codelineno-0-130" name="__codelineno-0-130"></a>    <span class="c1">#### Refit cooks if necessary ####</span>
</span><span id="__span-0-131"><a id="__codelineno-0-131" name="__codelineno-0-131"></a>    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">refit_cooks</span><span class="p">:</span>
</span><span id="__span-0-132"><a id="__codelineno-0-132" name="__codelineno-0-132"></a>        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Refitting Cook outliers...&quot;</span><span class="p">)</span>
</span><span id="__span-0-133"><a id="__codelineno-0-133" name="__codelineno-0-133"></a>        <span class="p">(</span>
</span><span id="__span-0-134"><a id="__codelineno-0-134" name="__codelineno-0-134"></a>            <span class="n">local_states</span><span class="p">,</span>
</span><span id="__span-0-135"><a id="__codelineno-0-135" name="__codelineno-0-135"></a>            <span class="n">gram_features_shared_states</span><span class="p">,</span>
</span><span id="__span-0-136"><a id="__codelineno-0-136" name="__codelineno-0-136"></a>            <span class="n">round_idx</span><span class="p">,</span>
</span><span id="__span-0-137"><a id="__codelineno-0-137" name="__codelineno-0-137"></a>        <span class="p">)</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">replace_outliers</span><span class="p">(</span>
</span><span id="__span-0-138"><a id="__codelineno-0-138" name="__codelineno-0-138"></a>            <span class="n">train_data_nodes</span><span class="p">,</span>
</span><span id="__span-0-139"><a id="__codelineno-0-139" name="__codelineno-0-139"></a>            <span class="n">aggregation_node</span><span class="p">,</span>
</span><span id="__span-0-140"><a id="__codelineno-0-140" name="__codelineno-0-140"></a>            <span class="n">local_states</span><span class="p">,</span>
</span><span id="__span-0-141"><a id="__codelineno-0-141" name="__codelineno-0-141"></a>            <span class="n">cooks_shared_state</span><span class="p">,</span>
</span><span id="__span-0-142"><a id="__codelineno-0-142" name="__codelineno-0-142"></a>            <span class="n">round_idx</span><span class="p">,</span>
</span><span id="__span-0-143"><a id="__codelineno-0-143" name="__codelineno-0-143"></a>            <span class="n">clean_models</span><span class="o">=</span><span class="n">clean_models</span><span class="p">,</span>
</span><span id="__span-0-144"><a id="__codelineno-0-144" name="__codelineno-0-144"></a>        <span class="p">)</span>
</span><span id="__span-0-145"><a id="__codelineno-0-145" name="__codelineno-0-145"></a>
</span><span id="__span-0-146"><a id="__codelineno-0-146" name="__codelineno-0-146"></a>        <span class="n">local_states</span><span class="p">,</span> <span class="n">round_idx</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">run_deseq2_lfc_dispersions</span><span class="p">(</span>
</span><span id="__span-0-147"><a id="__codelineno-0-147" name="__codelineno-0-147"></a>            <span class="n">train_data_nodes</span><span class="o">=</span><span class="n">train_data_nodes</span><span class="p">,</span>
</span><span id="__span-0-148"><a id="__codelineno-0-148" name="__codelineno-0-148"></a>            <span class="n">aggregation_node</span><span class="o">=</span><span class="n">aggregation_node</span><span class="p">,</span>
</span><span id="__span-0-149"><a id="__codelineno-0-149" name="__codelineno-0-149"></a>            <span class="n">local_states</span><span class="o">=</span><span class="n">local_states</span><span class="p">,</span>
</span><span id="__span-0-150"><a id="__codelineno-0-150" name="__codelineno-0-150"></a>            <span class="n">gram_features_shared_states</span><span class="o">=</span><span class="n">gram_features_shared_states</span><span class="p">,</span>
</span><span id="__span-0-151"><a id="__codelineno-0-151" name="__codelineno-0-151"></a>            <span class="n">round_idx</span><span class="o">=</span><span class="n">round_idx</span><span class="p">,</span>
</span><span id="__span-0-152"><a id="__codelineno-0-152" name="__codelineno-0-152"></a>            <span class="n">clean_models</span><span class="o">=</span><span class="n">clean_models</span><span class="p">,</span>
</span><span id="__span-0-153"><a id="__codelineno-0-153" name="__codelineno-0-153"></a>            <span class="n">refit_mode</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
</span><span id="__span-0-154"><a id="__codelineno-0-154" name="__codelineno-0-154"></a>        <span class="p">)</span>
</span><span id="__span-0-155"><a id="__codelineno-0-155" name="__codelineno-0-155"></a>        <span class="c1"># Replace values in the main ``local_adata`` object</span>
</span><span id="__span-0-156"><a id="__codelineno-0-156" name="__codelineno-0-156"></a>        <span class="n">local_states</span><span class="p">,</span> <span class="n">round_idx</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">replace_refitted_values</span><span class="p">(</span>
</span><span id="__span-0-157"><a id="__codelineno-0-157" name="__codelineno-0-157"></a>            <span class="n">train_data_nodes</span><span class="o">=</span><span class="n">train_data_nodes</span><span class="p">,</span>
</span><span id="__span-0-158"><a id="__codelineno-0-158" name="__codelineno-0-158"></a>            <span class="n">aggregation_node</span><span class="o">=</span><span class="n">aggregation_node</span><span class="p">,</span>
</span><span id="__span-0-159"><a id="__codelineno-0-159" name="__codelineno-0-159"></a>            <span class="n">local_states</span><span class="o">=</span><span class="n">local_states</span><span class="p">,</span>
</span><span id="__span-0-160"><a id="__codelineno-0-160" name="__codelineno-0-160"></a>            <span class="n">round_idx</span><span class="o">=</span><span class="n">round_idx</span><span class="p">,</span>
</span><span id="__span-0-161"><a id="__codelineno-0-161" name="__codelineno-0-161"></a>            <span class="n">clean_models</span><span class="o">=</span><span class="n">clean_models</span><span class="p">,</span>
</span><span id="__span-0-162"><a id="__codelineno-0-162" name="__codelineno-0-162"></a>        <span class="p">)</span>
</span><span id="__span-0-163"><a id="__codelineno-0-163" name="__codelineno-0-163"></a>
</span><span id="__span-0-164"><a id="__codelineno-0-164" name="__codelineno-0-164"></a>        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Finished refitting Cook outliers.&quot;</span><span class="p">)</span>
</span><span id="__span-0-165"><a id="__codelineno-0-165" name="__codelineno-0-165"></a>
</span><span id="__span-0-166"><a id="__codelineno-0-166" name="__codelineno-0-166"></a>    <span class="c1">#### Compute DESeq2 statistics ####</span>
</span><span id="__span-0-167"><a id="__codelineno-0-167" name="__codelineno-0-167"></a>
</span><span id="__span-0-168"><a id="__codelineno-0-168" name="__codelineno-0-168"></a>    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Running DESeq2 statistics.&quot;</span><span class="p">)</span>
</span><span id="__span-0-169"><a id="__codelineno-0-169" name="__codelineno-0-169"></a>
</span><span id="__span-0-170"><a id="__codelineno-0-170" name="__codelineno-0-170"></a>    <span class="n">local_states</span><span class="p">,</span> <span class="n">round_idx</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">run_deseq2_stats</span><span class="p">(</span>
</span><span id="__span-0-171"><a id="__codelineno-0-171" name="__codelineno-0-171"></a>        <span class="n">train_data_nodes</span><span class="p">,</span>
</span><span id="__span-0-172"><a id="__codelineno-0-172" name="__codelineno-0-172"></a>        <span class="n">aggregation_node</span><span class="p">,</span>
</span><span id="__span-0-173"><a id="__codelineno-0-173" name="__codelineno-0-173"></a>        <span class="n">local_states</span><span class="p">,</span>
</span><span id="__span-0-174"><a id="__codelineno-0-174" name="__codelineno-0-174"></a>        <span class="n">round_idx</span><span class="p">,</span>
</span><span id="__span-0-175"><a id="__codelineno-0-175" name="__codelineno-0-175"></a>        <span class="n">clean_models</span><span class="o">=</span><span class="n">clean_models</span><span class="p">,</span>
</span><span id="__span-0-176"><a id="__codelineno-0-176" name="__codelineno-0-176"></a>    <span class="p">)</span>
</span><span id="__span-0-177"><a id="__codelineno-0-177" name="__codelineno-0-177"></a>
</span><span id="__span-0-178"><a id="__codelineno-0-178" name="__codelineno-0-178"></a>    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Finished running DESeq2 statistics.&quot;</span><span class="p">)</span>
</span><span id="__span-0-179"><a id="__codelineno-0-179" name="__codelineno-0-179"></a>
</span><span id="__span-0-180"><a id="__codelineno-0-180" name="__codelineno-0-180"></a>    <span class="c1"># Build the results that will be downloaded at the end of the pipeline.</span>
</span><span id="__span-0-181"><a id="__codelineno-0-181" name="__codelineno-0-181"></a>
</span><span id="__span-0-182"><a id="__codelineno-0-182" name="__codelineno-0-182"></a>    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Saving pipeline results.&quot;</span><span class="p">)</span>
</span><span id="__span-0-183"><a id="__codelineno-0-183" name="__codelineno-0-183"></a>    <span class="bp">self</span><span class="o">.</span><span class="n">save_pipeline_results</span><span class="p">(</span>
</span><span id="__span-0-184"><a id="__codelineno-0-184" name="__codelineno-0-184"></a>        <span class="n">train_data_nodes</span><span class="p">,</span>
</span><span id="__span-0-185"><a id="__codelineno-0-185" name="__codelineno-0-185"></a>        <span class="n">aggregation_node</span><span class="p">,</span>
</span><span id="__span-0-186"><a id="__codelineno-0-186" name="__codelineno-0-186"></a>        <span class="n">local_states</span><span class="p">,</span>
</span><span id="__span-0-187"><a id="__codelineno-0-187" name="__codelineno-0-187"></a>        <span class="n">round_idx</span><span class="p">,</span>
</span><span id="__span-0-188"><a id="__codelineno-0-188" name="__codelineno-0-188"></a>        <span class="n">clean_models</span><span class="o">=</span><span class="n">clean_models</span><span class="p">,</span>
</span><span id="__span-0-189"><a id="__codelineno-0-189" name="__codelineno-0-189"></a>    <span class="p">)</span>
</span><span id="__span-0-190"><a id="__codelineno-0-190" name="__codelineno-0-190"></a>
</span><span id="__span-0-191"><a id="__codelineno-0-191" name="__codelineno-0-191"></a>    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Finished saving pipeline results.&quot;</span><span class="p">)</span>
</span></code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>




  </div>

    </div>

</div><h1 id="steps-of-fedpydeseq2-and-corresponding-mixin-classes">Steps of fedpydeseq2 and corresponding Mixin classes.</h1>
<h2 id="building-the-design-matrix-build_design_matrix">Building the design matrix: <a href="../build_design_matrix/build_design_matrix/">build_design_matrix</a></h2>
<h2 id="computing-the-size-factors-compute_size_factors">Computing the size factors: <a href="../compute_size_factors/compute_size_factors/">compute_size_factors</a></h2>
<h2 id="computing-log-fold-changes-and-dispersions-deseq2_lfc_dispersions">Computing log fold changes and dispersions: <a href="../deseq2_lfc_dispersions/deseq2_lfc_dispersions/">deseq2_lfc_dispersions</a></h2>
<h2 id="computing-the-cooks-distance-compute_cook_distance">Computing the cooks distance: <a href="../compute_cook_distance/compute_cook_distance/">compute_cook_distance</a></h2>
<h2 id="replacing-outliers-replace_outliers">Replacing outliers: <a href="../replace_outliers/replace_outliers/">replace_outliers</a></h2>
<h2 id="replace-refitted-values-replace_refitted_values">Replace refitted values: <a href="../replace_refitted_values/replace_refitted_values/">replace_refitted_values</a></h2>
<h2 id="computing-statistics-deseq2_stats">Computing statistics: <a href="../deseq2_stats/deseq2_stats/">deseq2_stats</a></h2>
<h2 id="saving-the-results-save_pipeline_results">Saving the results: <a href="../save_pipeline_results/save_pipeline_results/">save_pipeline_results</a></h2>
<h1 id="table-with-shared-quantities-between-centers-and-server">Table with shared quantities between centers and server</h1>
<table>
<thead>
<tr>
<th>ID</th>
<th>Name</th>
<th>Type</th>
<th>Shape</th>
<th>Description</th>
<th>Computed by</th>
<th>Sent to</th>
</tr>
</thead>
<tbody>
<tr>
<td>1</td>
<td>local_levels</td>
<td>dict</td>
<td></td>
<td>A dictionary whose keys are the names of the categorical factors, and whose values are the list of values taken by this factor in a given center. For example <span class="arithmatex">\(\texttt{\{stage: [Advanced, Non-advanced], gender: [female]\}}\)</span>.</td>
<td>Each center</td>
<td>Server</td>
</tr>
<tr>
<td>2</td>
<td>merged_levels</td>
<td>dict</td>
<td></td>
<td>A dictionary whose keys are the names of the categorical factors and whose values are arrays containing the list of values taken by this factor across all centers. For example <span class="arithmatex">\(\texttt{\{stage: [Advanced , Non-advanced], gender: [female, male]\}}\)</span>.</td>
<td>Server</td>
<td>Center</td>
</tr>
<tr>
<td>3</td>
<td>design_columns</td>
<td>Index</td>
<td><span class="arithmatex">\((p,)\)</span></td>
<td>The name of the columns of the local design matrix, before aggregation. They are of the form intercept, factor for continuous factors, and factor_level_vs_factor_ref_level otherwise. For example, <span class="arithmatex">\(\texttt{[intercept, stage\_Advanced\_vs\_Non-advanced, gender\_male\_vs\_female]}\)</span>.</td>
<td>Each center</td>
<td>Server</td>
</tr>
<tr>
<td>4</td>
<td>merged_columns</td>
<td>Index</td>
<td><span class="arithmatex">\((p,)\)</span></td>
<td>The union of the design columns across all centers. Local design matrices will then be updated to include all columns.</td>
<td>Server</td>
<td>Center</td>
</tr>
<tr>
<td>4</td>
<td>contrast</td>
<td>list</td>
<td></td>
<td>A list of three strings representing the contrast of interest, in case it is not specified by the user. Of the form <span class="arithmatex">\(\texttt{[factor, level1, level2]}\)</span>. For example, <span class="arithmatex">\(\texttt{[stage, Advanced, Non-advanced]}\)</span>.</td>
<td>Server</td>
<td>Center</td>
</tr>
<tr>
<td>5</td>
<td>log_mean</td>
<td>nparray</td>
<td><span class="arithmatex">\((G,)\)</span></td>
<td>For each gene, the mean of the log of the counts across all samples in a center <span class="arithmatex">\(\overline{\log(Y)}^{(k)}_{g} = \tfrac{1}{n_k}\sum_{i=1}^{n_k}\log(Y^{(k)}_{ig})\)</span>.</td>
<td>Each center</td>
<td>Server</td>
</tr>
<tr>
<td>5</td>
<td>n_samples</td>
<td>int</td>
<td></td>
<td>The number of samples in a center <span class="arithmatex">\(n_k\)</span> for each center <span class="arithmatex">\(k\)</span>.</td>
<td>Each center</td>
<td>Server</td>
</tr>
<tr>
<td>6</td>
<td>global_log_mean</td>
<td>nparray</td>
<td><span class="arithmatex">\((G,)\)</span></td>
<td>The mean of the log of the counts across all samples in all centers <span class="arithmatex">\(\overline{\log(Y)}_g =\sum_{k=1}^{K}\tfrac{n_k}{n}\overline{\log(Y)}^{(k)}_{g}\)</span>.</td>
<td>Server</td>
<td>Center</td>
</tr>
<tr>
<td>7</td>
<td>local_features</td>
<td>nparray</td>
<td><span class="arithmatex">\((p, G)\)</span></td>
<td><span class="arithmatex">\(\Phi^{(k)}  := X^{(k)\top} Z^{(k)}\)</span>.</td>
<td>Each center</td>
<td>Server</td>
</tr>
<tr>
<td>7</td>
<td>local_gram_matrix</td>
<td>nparray</td>
<td><span class="arithmatex">\((p, p)\)</span></td>
<td>The gram matrix of the local design matrix <span class="arithmatex">\(G^{(k)} := (X^{(k)})^{\top}X^{(k)}\)</span>.</td>
<td>Each center</td>
<td>Server</td>
</tr>
<tr>
<td>8</td>
<td>genewise_dispersions</td>
<td>nparray</td>
<td><span class="arithmatex">\((G,)\)</span></td>
<td>For each gene <span class="arithmatex">\(g\)</span>, the current estimate of the dispersion parameter <span class="arithmatex">\(\alpha_g\)</span>. This estimate is computed by first computing the global nll (summing all local nlls)  as well as the global Cox-Reid regularization term, which is half the log determinant of the sum of the local Cox-Reid matrices. The Cox-Reid regularized nll per gene and per dispersion in the grid is obtained by summing the regularization term and the nll. Finally, for every gene, the dispersion parameter is estimated by taking the minizer of this regularized nll on the grid of size <span class="arithmatex">\(N_{\texttt{gs}}\)</span>.</td>
<td>Server</td>
<td>Center</td>
</tr>
<tr>
<td>8</td>
<td>lower_log_bounds</td>
<td>nparray</td>
<td><span class="arithmatex">\((G,)\)</span></td>
<td>For each gene <span class="arithmatex">\(g\)</span>, the maximum of the log of the min dispersion and <span class="arithmatex">\(\alpha_g - \delta\)</span> where <span class="arithmatex">\(\delta\)</span> is the current mesh size of the grid and <span class="arithmatex">\(\alpha_g\)</span> is the current dispersion estimate. This value will be used as a lower bound for the next grid search for the dispersion parameter.</td>
<td>Server</td>
<td>Center</td>
</tr>
<tr>
<td>8</td>
<td>upper_log_bounds</td>
<td>nparray</td>
<td><span class="arithmatex">\((G,)\)</span></td>
<td>For each gene <span class="arithmatex">\(g\)</span>, the minimum of the log of the max dispersion and  <span class="arithmatex">\(\alpha_g + \delta\)</span> where <span class="arithmatex">\(\delta\)</span> is the current mesh size of the grid and <span class="arithmatex">\(\alpha_g\)</span> is the current dispersion estimate. This value will be used as an upper bound for the next grid search for the dispersion parameter.</td>
<td>Server</td>
<td>Center</td>
</tr>
<tr>
<td>9</td>
<td>disp_function_type</td>
<td>str</td>
<td></td>
<td>The type of the dispersion function used to model the dispersions. It can be either "parametric", if the iterative scheme to fit the trend curve has converged, the LBFGS-B method used to fit the parameters has converged, and the coefficienst of the trend curve are non-negative; or "mean" otherwise.</td>
<td>Server</td>
<td>Center</td>
</tr>
<tr>
<td>9</td>
<td>prior_disp_var</td>
<td>float</td>
<td><span class="arithmatex">\(()\)</span></td>
<td>A prior on the variance of the log dispersions around the log trend curve <span class="arithmatex">\(\sigma_{\texttt{trend}}^2\)</span>, estimated as the maximum between <span class="arithmatex">\(0.25\)</span> and <span class="arithmatex">\(\texttt{\_squared\_log\_res} - \psi_1((n-p)/2)\)</span>, where <span class="arithmatex">\(\psi_1(f/2)\)</span> is the variance of the log of a <span class="arithmatex">\(\chi^2_f\)</span> distribution. For more details, see \citep{love2014deseq2}.</td>
<td>Server</td>
<td>Center</td>
</tr>
<tr>
<td>9</td>
<td>_squared_logres</td>
<td>float</td>
<td><span class="arithmatex">\(()\)</span></td>
<td>The squared mean absolute deviation of the difference between the log of the genewise dispersions and the log of the fitted dispersions, restricted to the non-zero genes whose gene-wise dispersions are above <span class="arithmatex">\(100\times \texttt{min\_disp}\)</span>. The mean absolute deviation estimate is defined as the median of the absolute difference between the log residual and its median, scaled by the percent point function of the normal distribution at <span class="arithmatex">\(0.75\)</span>.</td>
<td>Server</td>
<td>Center</td>
</tr>
<tr>
<td>9</td>
<td>trend_coeffs</td>
<td>nparray</td>
<td><span class="arithmatex">\((2,)\)</span></td>
<td>The coefficients of the trend curve fitted to the dispersions. We model the dispersions <span class="arithmatex">\(\alpha_g\)</span> of the gene <span class="arithmatex">\(g\)</span> as a sample from an exponential distribution whose mean <span class="arithmatex">\(\alpha_{\texttt{trend}}(\overline{Z}_g)\)</span> is a function of <span class="arithmatex">\(\overline{Z}_g\)</span> parametrized by two coefficients <span class="arithmatex">\(\alpha_0\)</span> and <span class="arithmatex">\(\text{a}_1\)</span>; <span class="arithmatex">\(\alpha_{\texttt{trend}}(\overline{Z}_g) = \alpha_0 + \tfrac{\text{a}_1}{\overline{Z}_g}\)</span>. <span class="arithmatex">\(\alpha_{\texttt{trend}}\)</span> is called the trend curve. The coefficients are <span class="arithmatex">\((\alpha_0, \text{a}_1)\)</span>, and are obtained by starting with the set of non zero genes  and by iteratively i) minimizing the negative log likelihood of the exponential distribution on the set of genes with LBFGS-B and ii) removing the genes where the ratio between the dispersion <span class="arithmatex">\(\alpha_g\)</span> and the trend curve <span class="arithmatex">\(\alpha_{\texttt{trend}}(\overline{Z}_g)\)</span> is above <span class="arithmatex">\(15\)</span> or below <span class="arithmatex">\(10^{-4}\)</span> (we consider these genes as outliers to this model), until the set of genes is stable (see \citep{love2014deseq2} for more details).</td>
<td>Server</td>
<td>Center</td>
</tr>
<tr>
<td>9</td>
<td>fitted_dispersions</td>
<td>nparray</td>
<td><span class="arithmatex">\((G,)\)</span></td>
<td>For each gene <span class="arithmatex">\(g\)</span> with zero counts across all centers, <span class="arithmatex">\(\texttt{nan}\)</span>. For each non-zero gene <span class="arithmatex">\(g\)</span>, either <span class="arithmatex">\(\alpha_{\texttt{trend}}(\overline{Z}_g)\)</span> if the disp_function_type is "parametric" (i.e., the fitting of the parameters has converged), or <span class="arithmatex">\(\texttt{mean\_disp}\)</span> otherwise (i.e., if the disp_function_type is "mean"). Denoted with <span class="arithmatex">\(\alpha^{\texttt{trend}}_g\)</span>.</td>
<td>Server</td>
<td>Center</td>
</tr>
<tr>
<td>9</td>
<td>mean_disp</td>
<td>NoneType</td>
<td></td>
<td>None if the dispersion function type is "parametric", and the trimmed mean of the genewise dispersions whose value is above <span class="arithmatex">\(10\times \texttt{min\_disp}\)</span>, with trimming proportion <span class="arithmatex">\(0.001\)</span> otherwise.</td>
<td>Server</td>
<td>Center</td>
</tr>
<tr>
<td>12</td>
<td>trimmed_mean_normed_counts</td>
<td>DataFrame</td>
<td><span class="arithmatex">\((G, \|\mathcal{L}_{\geq 3}\|)\)</span></td>
<td>For each gene <span class="arithmatex">\(g\)</span> and each level <span class="arithmatex">\(l \in \mathcal{L}_{\geq 3}\)</span>, the corresponding entry <span class="arithmatex">\(\overline{Z}^{\texttt{trim}}_{g,l}\)</span> is the approximation of the trimmed mean of the normed counts for gene <span class="arithmatex">\(g\)</span> and samples whose line in the design corresponds to level <span class="arithmatex">\(l\)</span> with trim ratio <span class="arithmatex">\(r_{\texttt{trim}}\)</span>, computed by summing the trimmed_local_sums and dividing by the sum of the local n_samples for the corresponding gene and level.</td>
<td>Server</td>
<td>Center</td>
</tr>
<tr>
<td>14</td>
<td>varEst</td>
<td>nparray</td>
<td><span class="arithmatex">\((G,)\)</span></td>
<td>For each gene <span class="arithmatex">\(g\)</span>, the trimmed variance estimate of the normed counts, denoted with <span class="arithmatex">\(V^{\texttt{trim}}_g\)</span>. If use_lvl is <span class="arithmatex">\(\texttt{True}\)</span>, for each gene <span class="arithmatex">\(g\)</span>, it is the maximum across all admissible levels of the trimmed mean of the squared error for the gene and level in question, with trim ratio <span class="arithmatex">\(r_{\texttt{trim}}\)</span>, scaled (multiplied) by the scale factor <span class="arithmatex">\(1.51\)</span>. If use_level is <span class="arithmatex">\(\texttt{False}\)</span>, the trimmed mean of the squared error scaled by <span class="arithmatex">\(1.51\)</span> with trim ratio <span class="arithmatex">\(r_{\texttt{trim}}\)</span>.</td>
<td>Server</td>
<td>Center</td>
</tr>
<tr>
<td>15</td>
<td>local_hat_matrix</td>
<td>nparray</td>
<td><span class="arithmatex">\((G_{\texttt{nz}}, p, p)\)</span></td>
<td>For each gene <span class="arithmatex">\(g\)</span>, the hat matrix  <span class="arithmatex">\(H^{(k)}_{g} = (X^{(k)})^{\top} W^{(k)}_g X^{(k)}\)</span> where <span class="arithmatex">\(W^{(k)}_g \in \mathbb{R}^{n_k \times n_k}\)</span> is the diagonal matrix with diagonal entries <span class="arithmatex">\(\frac{\mu^{(k)}_{ig}}{1 + \mu^{(k)}_{ig} \alpha_g}\)</span> for <span class="arithmatex">\(1 \leq i \leq n_k\)</span>. <span class="arithmatex">\(\alpha_g\)</span> is the dispersion estimate of the gene and <span class="arithmatex">\(\mu^{(k)}_{ig}\)</span> is the expected value of the gene for sample <span class="arithmatex">\(i\)</span> for parameter <span class="arithmatex">\(\beta\)</span>, that is <span class="arithmatex">\(\gamma^{(k)}_{g} \exp(X^{(k)}_{i} \cdot \beta_g)\)</span>.</td>
<td>Each center</td>
<td>Server</td>
</tr>
<tr>
<td>15</td>
<td>mean_normed_counts</td>
<td>nparray</td>
<td><span class="arithmatex">\((G,)\)</span></td>
<td>For each gene, the mean of the local normed counts, i.e., <span class="arithmatex">\(\overline{Z}^{(k)}_{g} = \tfrac{1}{n_k}\sum_{i=1}^{n_k}{Z^{(k)}_{ig}}\)</span>.</td>
<td>Each center</td>
<td>Server</td>
</tr>
<tr>
<td>15</td>
<td>n_samples</td>
<td>int</td>
<td></td>
<td>The number of samples in a center <span class="arithmatex">\(n_k\)</span> for each center <span class="arithmatex">\(k\)</span>.</td>
<td>Each center</td>
<td>Server</td>
</tr>
<tr>
<td>15</td>
<td>varEst</td>
<td>nparray</td>
<td><span class="arithmatex">\((G,)\)</span></td>
<td>For each gene <span class="arithmatex">\(g\)</span>, the trimmed variance estimate of the normed counts, denoted with <span class="arithmatex">\(V^{\texttt{trim}}_g\)</span>. This quantitiy is passed on from the previous shared state.</td>
<td>Each center</td>
<td>Server</td>
</tr>
<tr>
<td>15</td>
<td>_skip_cooks</td>
<td>bool</td>
<td></td>
<td>A boolean indicating whether to skip the computation of the intermediate quantities to compute the  of the Cook's distance. This is set to <span class="arithmatex">\(\texttt{True}\)</span> if the Cook's distance is stored in the local state (which is not the case by default due to memory issues). Otherwise, it is set to <span class="arithmatex">\(\texttt{False}\)</span> (default behaviour).</td>
<td>Each center</td>
<td>Server</td>
</tr>
<tr>
<td>16</td>
<td>cooks_dispersions</td>
<td>nparray</td>
<td><span class="arithmatex">\((G,)\)</span></td>
<td>For each gene <span class="arithmatex">\(g\)</span>, a robust estimate of the dispersion parameter <span class="arithmatex">\(\alpha^{\texttt{cooks}}_g\)</span> computed from the trimmed variance estimate and the global mean of the normed counts. We compute this estimate as <span class="arithmatex">\(\max((V^{\texttt{trim}}_g- \overline{Y}_g)/\overline{Y}_g^2,0.04)\)</span>.</td>
<td>Server</td>
<td>Center</td>
</tr>
<tr>
<td>16</td>
<td>global_hat_matrix_inv</td>
<td>nparray</td>
<td><span class="arithmatex">\((G_{\texttt{nz}}, p, p)\)</span></td>
<td>For each gene <span class="arithmatex">\(g\)</span>, we compute the global hat matrix as the sum of the local hat matrices, and its inverse.</td>
<td>Server</td>
<td>Center</td>
</tr>
<tr>
<td>17</td>
<td>replaceable_samples</td>
<td>bool</td>
<td><span class="arithmatex">\(()\)</span></td>
<td>A boolean indicating if there are any replaceable samples in the center. A sample <span class="arithmatex">\(i\)</span> of center <span class="arithmatex">\(k\)</span> is said to be replaceable if there are at least <span class="arithmatex">\(\texttt{min\_replicates}\)</span> samples across all centers which share the same design factor levels as <span class="arithmatex">\(i\)</span>. <span class="arithmatex">\(\texttt{min\_replicates}\)</span> is a user defined parameter, set to <span class="arithmatex">\(7\)</span> by default.</td>
<td>Each center</td>
<td>Server</td>
</tr>
<tr>
<td>17</td>
<td>local_genes_to_replace</td>
<td>set</td>
<td></td>
<td>The set of genes <span class="arithmatex">\(g\)</span> for which the Cook's distance is above the cutoff value for any sample in the center. For a given gene <span class="arithmatex">\(g\)</span> and given sample <span class="arithmatex">\(i\)</span>, the Cook's distance is computed as <span class="arithmatex">\(\tfrac{h^{(k)}_{ig}}{p (1-h^{(k)}_{ig})^2}~(R^2)^{(k)}_{ig}\)</span> where <span class="arithmatex">\((R^2)^{(k)}_{ig}\)</span> is the squared Pearson residual of the negative binomial GLM with log fold changes <span class="arithmatex">\(\beta_g\)</span> and dispersions <span class="arithmatex">\(\alpha^{\texttt{cooks}}_g\)</span>, computed as <span class="arithmatex">\((Y^{(k)}_{ig} - \mu^{(k)}_{ig})^2/(V^{\texttt{NB}})^{(k)}_{ig}\)</span>, and <span class="arithmatex">\(h^{(k)}_{ig}\)</span> is the <span class="arithmatex">\(i\)</span>-th diagonal element of <span class="arithmatex">\(X^{(k)} H^{-1}_{g} (X^{(k)})^{\top}\)</span>, where <span class="arithmatex">\(H^{-1}_{g}\)</span> is the inverse of the global hat matrix. The cutoff value is set to the <span class="arithmatex">\(0.99\)</span>-th quantile of the F-distribution with <span class="arithmatex">\(p\)</span> and <span class="arithmatex">\(n-p\)</span> degrees of freedom. Here, <span class="arithmatex">\(\mu^{(k)}_{ig} = \gamma^{(k)}_{i}~\exp(X^{(k)}_{i} \cdot \beta_g)\)</span> and <span class="arithmatex">\((V^{\texttt{NB}})^{(k)}_{ig} = \mu^{(k)}_{ig} (1 + \mu^{(k)}_{ig} \alpha^{\texttt{cooks}}_g)\)</span>.</td>
<td>Each center</td>
<td>Server</td>
</tr>
<tr>
<td>18</td>
<td>genes_to_replace</td>
<td>set</td>
<td></td>
<td>The set of genes <span class="arithmatex">\(g\)</span> for which the Cook's distance is above the cutoff value for any sample in any center (the union of the local genes to replace across all centers).</td>
<td>Server</td>
<td>Center</td>
</tr>
<tr>
<td>20</td>
<td>trimmed_mean_normed_counts</td>
<td>nparray</td>
<td><span class="arithmatex">\((G_{\texttt{r}},)\)</span></td>
<td>For each gene to replace <span class="arithmatex">\(g\)</span>, the trimmed mean of the normed counts across all samples, denoted with <span class="arithmatex">\(\overline{Z}^{\texttt{trim}}_{g}\)</span>, with trim ratio set to <span class="arithmatex">\(0.2\)</span>.</td>
<td>Server</td>
<td>Center</td>
</tr>
<tr>
<td>21</td>
<td>loc_new_all_zeroes</td>
<td>nparray</td>
<td><span class="arithmatex">\((G_{\texttt{r}},)\)</span></td>
<td>A boolean array which for each gene to replace <span class="arithmatex">\(g\)</span> indicates if the new count matrix of the center is all zeroes (across samples) for this gene (the new count matrix is computed by imputing the Cook's outliers with the trimmed mean of the normed counts times the size factor).</td>
<td>Each center</td>
<td>Server</td>
</tr>
<tr>
<td>22</td>
<td>new_all_zeroes</td>
<td>nparray</td>
<td><span class="arithmatex">\((G_{\texttt{r}},)\)</span></td>
<td>A boolean array which for each gene to replace <span class="arithmatex">\(g\)</span> indicates if all counts across all centers are zero for this gene.</td>
<td>Server</td>
<td>Center</td>
</tr>
<tr>
<td>23</td>
<td>local_features</td>
<td>nparray</td>
<td><span class="arithmatex">\((p, G_{\texttt{act}})\)</span></td>
<td><span class="arithmatex">\(\Phi^{(k)}  := (X^{(k)})^{\top} Z^{(k)}\)</span>, where <span class="arithmatex">\(Z^{(k)}\)</span> is the normalized counts in the center <span class="arithmatex">\(k\)</span> on the set of genes to replace, where the value of Cook's outliers have been replaced using a trimmed mean.</td>
<td>Each center</td>
<td>Server</td>
</tr>
<tr>
<td>24</td>
<td>genewise_dispersions</td>
<td>nparray</td>
<td><span class="arithmatex">\((G_{\texttt{r}},)\)</span></td>
<td>For each gene <span class="arithmatex">\(g\)</span>, the current estimate of the dispersion parameter <span class="arithmatex">\(\alpha_g\)</span>. This estimate is computed by first computing the global nll (summing all local nlls)  as well as the global Cox-Reid regularization term, which is half the log determinant of the sum of the local Cox-Reid matrices. The Cox-Reid regularized nll per gene and per dispersion in the grid is obtained by summing the regularization term and the nll. Finally, for every gene, the dispersion parameter is estimated by taking the minizer of this regularized nll on the grid of size <span class="arithmatex">\(N_{\texttt{gs}}\)</span>.</td>
<td>Server</td>
<td>Center</td>
</tr>
<tr>
<td>24</td>
<td>lower_log_bounds</td>
<td>nparray</td>
<td><span class="arithmatex">\((G_{\texttt{r}},)\)</span></td>
<td>For each gene <span class="arithmatex">\(g\)</span>, the maximum of the log of the min dispersion and <span class="arithmatex">\(\alpha_g - \delta\)</span> where <span class="arithmatex">\(\delta\)</span> is the current mesh size of the grid and <span class="arithmatex">\(\alpha_g\)</span> is the current dispersion estimate. This value will be used as a lower bound for the next grid search for the dispersion parameter.</td>
<td>Server</td>
<td>Center</td>
</tr>
<tr>
<td>24</td>
<td>upper_log_bounds</td>
<td>nparray</td>
<td><span class="arithmatex">\((G_{\texttt{r}},)\)</span></td>
<td>For each gene <span class="arithmatex">\(g\)</span>, the minimum of the log of the max dispersion and  <span class="arithmatex">\(\alpha_g + \delta\)</span> where <span class="arithmatex">\(\delta\)</span> is the current mesh size of the grid and <span class="arithmatex">\(\alpha_g\)</span> is the current dispersion estimate. This value will be used as an upper bound for the next grid search for the dispersion parameter.</td>
<td>Server</td>
<td>Center</td>
</tr>
<tr>
<td>29</td>
<td>global_hat_matrix_inv</td>
<td>nparray</td>
<td><span class="arithmatex">\((G_{\texttt{nz}}, p, p)\)</span></td>
<td>For each gene <span class="arithmatex">\(g\)</span>, we compute the global hat matrix as the sum of the local hat matrices, and its inverse.</td>
<td>Server</td>
<td>Center</td>
</tr>
<tr>
<td>29</td>
<td>cooks_dispersions</td>
<td>nparray</td>
<td><span class="arithmatex">\((G,)\)</span></td>
<td>For each gene <span class="arithmatex">\(g\)</span>, a robust estimate of the dispersion parameter <span class="arithmatex">\(\alpha^{\texttt{cooks}}_g\)</span> computed from the trimmed variance estimate and the global mean of the normed counts. We compute this estimate as <span class="arithmatex">\(\max((V^{\texttt{trim}}_g- \overline{Y}_g)/\overline{Y}_g^2,0.04)\)</span>.</td>
<td>Server</td>
<td>Center</td>
</tr>
<tr>
<td>29</td>
<td>p_values</td>
<td>nparray</td>
<td><span class="arithmatex">\((G,)\)</span></td>
<td>For each gene, the p-value of the Wald statistic, computed from the survival function of the normal distribution applied to the wald statistic.</td>
<td>Server</td>
<td>Center</td>
</tr>
<tr>
<td>29</td>
<td>wald_statistics</td>
<td>nparray</td>
<td><span class="arithmatex">\((G,)\)</span></td>
<td>For each gene, the Wald statistic of the gene expression. This statistics depends on the <span class="arithmatex">\(\texttt{lfc\_null}\)</span> parameter which sets the null hypothesis on the log fold change (set to <span class="arithmatex">\(0\)</span> by default), and the <span class="arithmatex">\(\texttt{alt\_hypothesis}\)</span> parameter, which defines the alternative hypothesis on the log fold change (set to <span class="arithmatex">\(\texttt{None}\)</span> by default, can be <span class="arithmatex">\(\texttt{greater}, ~\texttt{greaterAbs}, ~\texttt{less},~ \texttt{lessAbs}\)</span>). If the alternative hypothesis is <span class="arithmatex">\(\texttt{None}\)</span>, then the Wald statistic is computed as the centered normalized log fold change.</td>
<td>Server</td>
<td>Center</td>
</tr>
<tr>
<td>29</td>
<td>wald_se</td>
<td>nparray</td>
<td><span class="arithmatex">\((G,)\)</span></td>
<td>For each gene, the standard error on the log fold change value for the given contrast given by the GLM.</td>
<td>Server</td>
<td>Center</td>
</tr>
<tr>
<td>30</td>
<td>p_values</td>
<td>nparray</td>
<td><span class="arithmatex">\((G,)\)</span></td>
<td>If Cook's filtering is enabled (which is the case by default with the $\texttt{cooks_filter} $ parameter), then the p-values of genes which have <span class="arithmatex">\(\leq 2\)</span> samples above the gene count of the sample maximizing the Cook's distance across all centers are set to <span class="arithmatex">\(\texttt{nan}\)</span>. Otherwise, passed on without modification.</td>
<td>Server</td>
<td>Center</td>
</tr>
<tr>
<td>30</td>
<td>wald_statistics</td>
<td>nparray</td>
<td><span class="arithmatex">\((G,)\)</span></td>
<td>For each gene, the Wald statistic of the gene expression. Passed on without modification.</td>
<td>Server</td>
<td>Center</td>
</tr>
<tr>
<td>30</td>
<td>wald_se</td>
<td>nparray</td>
<td><span class="arithmatex">\((G,)\)</span></td>
<td>For each gene, the standard error on the log fold change value for the given contrast given by the GLM. Passed on without modification.</td>
<td>Server</td>
<td>Center</td>
</tr>
<tr>
<td>32</td>
<td>contrast</td>
<td>list</td>
<td></td>
<td>A list of three strings representing the contrast of interest, in case it is not specified by the user. Of the form <span class="arithmatex">\(\texttt{[factor, level1, level2]}\)</span>. For example, <span class="arithmatex">\(\texttt{[stage, Advanced, Non-advanced]}\)</span>.</td>
<td>Each center</td>
<td>Server</td>
</tr>
<tr>
<td>32</td>
<td>_squared_logres</td>
<td>float</td>
<td><span class="arithmatex">\(()\)</span></td>
<td>The squared mean absolute deviation of the difference between the log of the genewise dispersions and the log of the fitted dispersions, restricted to the non-zero genes whose gene-wise dispersions are above <span class="arithmatex">\(100\times \texttt{min\_disp}\)</span>. The mean absolute deviation estimate is defined as the median of the absolute difference between the log residual and its median, scaled by the percent point function of the normal distribution at <span class="arithmatex">\(0.75\)</span>.</td>
<td>Each center</td>
<td>Server</td>
</tr>
<tr>
<td>32</td>
<td>refitted</td>
<td>nparray</td>
<td><span class="arithmatex">\((G,)\)</span></td>
<td>A boolean array marking genes which can be replaced and which, after replacing the count value by the imputation value, are non-zero.</td>
<td>Each center</td>
<td>Server</td>
</tr>
<tr>
<td>32</td>
<td>replaced</td>
<td>nparray</td>
<td><span class="arithmatex">\((G,)\)</span></td>
<td>A boolean array marking genes <span class="arithmatex">\(g\)</span> for which the Cook's distance is above the cutoff value for any sample in any center (the union of the local genes to replace across all centers).</td>
<td>Each center</td>
<td>Server</td>
</tr>
<tr>
<td>32</td>
<td>wald_se</td>
<td>nparray</td>
<td><span class="arithmatex">\((G,)\)</span></td>
<td>For each gene, the standard error on the log fold change value for the given contrast given by the GLM.</td>
<td>Each center</td>
<td>Server</td>
</tr>
<tr>
<td>32</td>
<td>wald_statistics</td>
<td>nparray</td>
<td><span class="arithmatex">\((G,)\)</span></td>
<td>For each gene, the Wald statistic of the gene expression.</td>
<td>Each center</td>
<td>Server</td>
</tr>
<tr>
<td>32</td>
<td>p_values</td>
<td>nparray</td>
<td><span class="arithmatex">\((G,)\)</span></td>
<td>For each gene, the p-value of the Wald statistic, computed from the survival function of the normal distribution applied to the wald statistic, and set to <span class="arithmatex">\(\texttt{nan}\)</span> if the gene is a Cook's outlier if <span class="arithmatex">\(\texttt{cooks\_filter}\)</span> is enabled.</td>
<td>Each center</td>
<td>Server</td>
</tr>
<tr>
<td>32</td>
<td>prior_disp_var</td>
<td>float</td>
<td><span class="arithmatex">\(()\)</span></td>
<td>A prior on the variance of the log dispersions around the log trend curve <span class="arithmatex">\(\sigma_{\texttt{trend}}^2\)</span>, estimated as the maximum between <span class="arithmatex">\(0.25\)</span> and <span class="arithmatex">\(\texttt{\_squared\_log\_res} - \psi_1((n-p)/2)\)</span>, where <span class="arithmatex">\(\psi_1(f/2)\)</span> is the variance of the log of a <span class="arithmatex">\(\chi^2_f\)</span> distribution. For more details, see \citep{love2014deseq2}.</td>
<td>Each center</td>
<td>Server</td>
</tr>
<tr>
<td>32</td>
<td>LFC</td>
<td>DataFrame</td>
<td><span class="arithmatex">\((G, p)\)</span></td>
<td>The log fold changes of the gene expression. This dataframe is indexed by genes on one hand, and by design column names on the other.</td>
<td>Each center</td>
<td>Server</td>
</tr>
<tr>
<td>32</td>
<td>fitted_dispersions</td>
<td>nparray</td>
<td><span class="arithmatex">\((G,)\)</span></td>
<td>For each gene <span class="arithmatex">\(g\)</span> with zero counts across all centers, <span class="arithmatex">\(\texttt{nan}\)</span>. For each non-zero gene <span class="arithmatex">\(g\)</span>, either <span class="arithmatex">\(\alpha_{\texttt{trend}}(\overline{Z}_g)\)</span> if the disp_function_type is "parametric" (i.e., the fitting of the parameters has converged), or <span class="arithmatex">\(\texttt{mean\_disp}\)</span> otherwise (i.e., if the disp_function_type is "mean"). Denoted with <span class="arithmatex">\(\alpha^{\texttt{trend}}_g\)</span>.</td>
<td>Each center</td>
<td>Server</td>
</tr>
<tr>
<td>32</td>
<td>non_zero</td>
<td>nparray</td>
<td><span class="arithmatex">\((G,)\)</span></td>
<td>A boolean array indicating which genes have non-zero counts in at least one center.</td>
<td>Each center</td>
<td>Server</td>
</tr>
<tr>
<td>32</td>
<td>genewise_dispersions</td>
<td>nparray</td>
<td><span class="arithmatex">\((G,)\)</span></td>
<td>For each gene <span class="arithmatex">\(g\)</span>, the current estimate of the dispersion parameter <span class="arithmatex">\(\alpha_g\)</span>. This estimate is computed by first computing the global nll (summing all local nlls)  as well as the global Cox-Reid regularization term, which is half the log determinant of the sum of the local Cox-Reid matrices. The Cox-Reid regularized nll per gene and per dispersion in the grid is obtained by summing the regularization term and the nll. Finally, for every gene, the dispersion parameter is estimated by taking the minizer of this regularized nll on the grid of size <span class="arithmatex">\(N_{\texttt{gs}}\)</span>.</td>
<td>Each center</td>
<td>Server</td>
</tr>
<tr>
<td>32</td>
<td>dispersions</td>
<td>nparray</td>
<td><span class="arithmatex">\((G,)\)</span></td>
<td>The estimated dispersions for each gene, which are the MAP dispersions if the gene is not an outlier w.r.t. the trend curve, and the gene-wise dispersions otherwise.</td>
<td>Each center</td>
<td>Server</td>
</tr>
<tr>
<td>32</td>
<td>MAP_dispersions</td>
<td>nparray</td>
<td><span class="arithmatex">\((G,)\)</span></td>
<td>For each gene <span class="arithmatex">\(g\)</span>, the current estimate of the MAP dispersion parameter <span class="arithmatex">\(\alpha_g\)</span>. This estimate is computed by first computing the global nll (summing all local nlls)  as well as the global Cox-Reid regularization term, which is half the log determinant of the sum of the local Cox-Reid matrices, and the regularization coming from the prior on the log dispersions around the trend curve. The Cox-Reid, prior regularized nll per gene and per dispersion in the grid is obtained by summing the regularization terms and the nll. Finally, for every gene, the MAP dispersion parameter is estimated by taking the minizer of this regularized nll on the grid of size <span class="arithmatex">\(N_{\texttt{gs}}\)</span>.</td>
<td>Each center</td>
<td>Server</td>
</tr>
<tr>
<td>32</td>
<td>gene_names</td>
<td>Index</td>
<td><span class="arithmatex">\((G,)\)</span></td>
<td>The gene names.</td>
<td>Each center</td>
<td>Server</td>
</tr>
<tr>
<td>32</td>
<td>padj</td>
<td>Series</td>
<td><span class="arithmatex">\((G,)\)</span></td>
<td>The adjusted p-values for each gene, computed from the p-values using independent filtering if the <span class="arithmatex">\(\texttt{independent\_filter}\)</span> parameter is <span class="arithmatex">\(\texttt{True}\)</span> (default), and the Benjamini-Hochberg procedure otherwise.</td>
<td>Each center</td>
<td>Server</td>
</tr>
<tr>
<td>33</td>
<td>prior_disp_var</td>
<td>float</td>
<td><span class="arithmatex">\(()\)</span></td>
<td>A prior on the variance of the log dispersions around the log trend curve <span class="arithmatex">\(\sigma_{\texttt{trend}}^2\)</span>, estimated as the maximum between <span class="arithmatex">\(0.25\)</span> and <span class="arithmatex">\(\texttt{\_squared\_log\_res} - \psi_1((n-p)/2)\)</span>, where <span class="arithmatex">\(\psi_1(f/2)\)</span> is the variance of the log of a <span class="arithmatex">\(\chi^2_f\)</span> distribution. For more details, see \citep{love2014deseq2}.</td>
<td>Server</td>
<td>All</td>
</tr>
<tr>
<td>33</td>
<td>refitted</td>
<td>nparray</td>
<td><span class="arithmatex">\((G,)\)</span></td>
<td>A boolean array marking genes which can be replaced and which, after replacing the count value by the imputation value, are non-zero.</td>
<td>Server</td>
<td>All</td>
</tr>
<tr>
<td>33</td>
<td>replaced</td>
<td>nparray</td>
<td><span class="arithmatex">\((G,)\)</span></td>
<td>A boolean array marking genes <span class="arithmatex">\(g\)</span> for which the Cook's distance is above the cutoff value for any sample in any center (the union of the local genes to replace across all centers).</td>
<td>Server</td>
<td>All</td>
</tr>
<tr>
<td>33</td>
<td>wald_se</td>
<td>nparray</td>
<td><span class="arithmatex">\((G,)\)</span></td>
<td>For each gene, the standard error on the log fold change value for the given contrast given by the GLM.</td>
<td>Server</td>
<td>All</td>
</tr>
<tr>
<td>33</td>
<td>wald_statistics</td>
<td>nparray</td>
<td><span class="arithmatex">\((G,)\)</span></td>
<td>For each gene, the Wald statistic of the gene expression.</td>
<td>Server</td>
<td>All</td>
</tr>
<tr>
<td>33</td>
<td>p_values</td>
<td>nparray</td>
<td><span class="arithmatex">\((G,)\)</span></td>
<td>For each gene, the p-value of the Wald statistic, computed from the survival function of the normal distribution applied to the wald statistic, and set to <span class="arithmatex">\(\texttt{nan}\)</span> if the gene is a Cook's outlier if <span class="arithmatex">\(\texttt{cooks\_filter}\)</span> is enabled.</td>
<td>Server</td>
<td>All</td>
</tr>
<tr>
<td>33</td>
<td>padj</td>
<td>Series</td>
<td><span class="arithmatex">\((G,)\)</span></td>
<td>The adjusted p-values for each gene, computed from the p-values using independent filtering if the <span class="arithmatex">\(\texttt{independent\_filter}\)</span> parameter is <span class="arithmatex">\(\texttt{True}\)</span> (default), and the Benjamini-Hochberg procedure otherwise.</td>
<td>Server</td>
<td>All</td>
</tr>
<tr>
<td>33</td>
<td>genewise_dispersions</td>
<td>nparray</td>
<td><span class="arithmatex">\((G,)\)</span></td>
<td>For each gene <span class="arithmatex">\(g\)</span>, the current estimate of the dispersion parameter <span class="arithmatex">\(\alpha_g\)</span>. This estimate is computed by first computing the global nll (summing all local nlls)  as well as the global Cox-Reid regularization term, which is half the log determinant of the sum of the local Cox-Reid matrices. The Cox-Reid regularized nll per gene and per dispersion in the grid is obtained by summing the regularization term and the nll. Finally, for every gene, the dispersion parameter is estimated by taking the minizer of this regularized nll on the grid of size <span class="arithmatex">\(N_{\texttt{gs}}\)</span>.</td>
<td>Server</td>
<td>All</td>
</tr>
<tr>
<td>33</td>
<td>fitted_dispersions</td>
<td>nparray</td>
<td><span class="arithmatex">\((G,)\)</span></td>
<td>For each gene <span class="arithmatex">\(g\)</span> with zero counts across all centers, <span class="arithmatex">\(\texttt{nan}\)</span>. For each non-zero gene <span class="arithmatex">\(g\)</span>, either <span class="arithmatex">\(\alpha_{\texttt{trend}}(\overline{Z}_g)\)</span> if the disp_function_type is "parametric" (i.e., the fitting of the parameters has converged), or <span class="arithmatex">\(\texttt{mean\_disp}\)</span> otherwise (i.e., if the disp_function_type is "mean"). Denoted with <span class="arithmatex">\(\alpha^{\texttt{trend}}_g\)</span>.</td>
<td>Server</td>
<td>All</td>
</tr>
<tr>
<td>33</td>
<td>non_zero</td>
<td>nparray</td>
<td><span class="arithmatex">\((G,)\)</span></td>
<td>A boolean array indicating which genes have non-zero counts in at least one center.</td>
<td>Server</td>
<td>All</td>
</tr>
<tr>
<td>33</td>
<td>dispersions</td>
<td>nparray</td>
<td><span class="arithmatex">\((G,)\)</span></td>
<td>The estimated dispersions for each gene, which are the MAP dispersions if the gene is not an outlier w.r.t. the trend curve, and the gene-wise dispersions otherwise.</td>
<td>Server</td>
<td>All</td>
</tr>
<tr>
<td>33</td>
<td>MAP_dispersions</td>
<td>nparray</td>
<td><span class="arithmatex">\((G,)\)</span></td>
<td>For each gene <span class="arithmatex">\(g\)</span>, the current estimate of the MAP dispersion parameter <span class="arithmatex">\(\alpha_g\)</span>. This estimate is computed by first computing the global nll (summing all local nlls)  as well as the global Cox-Reid regularization term, which is half the log determinant of the sum of the local Cox-Reid matrices, and the regularization coming from the prior on the log dispersions around the trend curve. The Cox-Reid, prior regularized nll per gene and per dispersion in the grid is obtained by summing the regularization terms and the nll. Finally, for every gene, the MAP dispersion parameter is estimated by taking the minizer of this regularized nll on the grid of size <span class="arithmatex">\(N_{\texttt{gs}}\)</span>.</td>
<td>Server</td>
<td>All</td>
</tr>
<tr>
<td>33</td>
<td>gene_names</td>
<td>Index</td>
<td><span class="arithmatex">\((G,)\)</span></td>
<td>The gene names.</td>
<td>Server</td>
<td>All</td>
</tr>
<tr>
<td>33</td>
<td>_squared_logres</td>
<td>float</td>
<td><span class="arithmatex">\(()\)</span></td>
<td>The squared mean absolute deviation of the difference between the log of the genewise dispersions and the log of the fitted dispersions, restricted to the non-zero genes whose gene-wise dispersions are above <span class="arithmatex">\(100\times \texttt{min\_disp}\)</span>. The mean absolute deviation estimate is defined as the median of the absolute difference between the log residual and its median, scaled by the percent point function of the normal distribution at <span class="arithmatex">\(0.75\)</span>.</td>
<td>Server</td>
<td>All</td>
</tr>
<tr>
<td>33</td>
<td>LFC</td>
<td>DataFrame</td>
<td><span class="arithmatex">\((G, p)\)</span></td>
<td>The log fold changes of the gene expression. This dataframe is indexed by genes on one hand, and by design column names on the other.</td>
<td>Server</td>
<td>All</td>
</tr>
<tr>
<td>33</td>
<td>contrast</td>
<td>list</td>
<td></td>
<td>A list of three strings representing the contrast of interest, in case it is not specified by the user. Of the form <span class="arithmatex">\(\texttt{[factor, level1, level2]}\)</span>. For example, <span class="arithmatex">\(\texttt{[stage, Advanced, Non-advanced]}\)</span>.</td>
<td>Server</td>
<td>All</td>
</tr>
</tbody>
</table>







  
    
  
  
    
  


  <aside class="md-source-file">
    
      
  <span class="md-source-file__fact">
    <span class="md-icon" title="Last update">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M21 13.1c-.1 0-.3.1-.4.2l-1 1 2.1 2.1 1-1c.2-.2.2-.6 0-.8l-1.3-1.3c-.1-.1-.2-.2-.4-.2m-1.9 1.8-6.1 6V23h2.1l6.1-6.1zM12.5 7v5.2l4 2.4-1 1L11 13V7zM11 21.9c-5.1-.5-9-4.8-9-9.9C2 6.5 6.5 2 12 2c5.3 0 9.6 4.1 10 9.3-.3-.1-.6-.2-1-.2s-.7.1-1 .2C19.6 7.2 16.2 4 12 4c-4.4 0-8 3.6-8 8 0 4.1 3.1 7.5 7.1 7.9l-.1.2z"/></svg>
    </span>
    <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-timeago"><span class="timeago" datetime="2025-06-24T07:38:14+00:00" locale="en"></span></span><span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-iso_date">2025-06-24</span>
  </span>

    
    
      
  <span class="md-source-file__fact">
    <span class="md-icon" title="Created">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M14.47 15.08 11 13V7h1.5v5.25l3.08 1.83c-.41.28-.79.62-1.11 1m-1.39 4.84c-.36.05-.71.08-1.08.08-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8c0 .37-.03.72-.08 1.08.69.1 1.33.32 1.92.64.1-.56.16-1.13.16-1.72 0-5.5-4.5-10-10-10S2 6.5 2 12s4.47 10 10 10c.59 0 1.16-.06 1.72-.16-.32-.59-.54-1.23-.64-1.92M18 15v3h-3v2h3v3h2v-3h3v-2h-3v-3z"/></svg>
    </span>
    <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-timeago"><span class="timeago" datetime="2025-06-24T07:38:14+00:00" locale="en"></span></span><span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-iso_date">2025-06-24</span>
  </span>

    
    
    
  </aside>





                
              </article>
            </div>
          
          
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var labels=set.querySelector(".tabbed-labels");for(var tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Copyright (c) 2024-present Owkin Inc.
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
        <div class="md-social">
  
    
    
    
    
      
      
    
    <a href="https://github.com/owkin/fedpydeseq2" target="_blank" rel="noopener" title="github.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8M97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "../../../..", "features": ["content.code.copy", "content.tabs.link", "palette.switch", "navigation.indexes"], "search": "../../../../assets/javascripts/workers/search.f8cc74c7.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../../../../assets/javascripts/bundle.c8b220af.min.js"></script>
      
        <script src="../../../../js/timeago.min.js"></script>
      
        <script src="../../../../js/timeago_mkdocs_material.js"></script>
      
        <script src="../../../../javascripts/katex.js"></script>
      
        <script src="https://unpkg.com/katex@0/dist/katex.min.js"></script>
      
        <script src="https://unpkg.com/katex@0/dist/contrib/auto-render.min.js"></script>
      
    
  </body>
</html>